{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "70ec3133-ce7b-47ed-a7e5-7ebd83f3d215",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import pandas_plink as pdpl\n",
    "import statsmodels.api as sm\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "import subprocess\n",
    "from sklearn.metrics import r2_score\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ab28f33c-087f-4f8e-948b-1e63664a136e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Mapping files:   0%|          | 0/3 [00:00<?, ?it/s]\u001b[A\n",
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 26.65it/s]\u001b[A\n"
     ]
    }
   ],
   "source": [
    "# There are three file types for genetic data that interest us\n",
    "# Multiple different samples of these files can be found on 1000 Genomes\n",
    "# In the interest of size and reproducibility we will use just EUR.14 to showcase the method\n",
    "# This method can be reproduced with any of the chromosomes included in the data folder\n",
    "bim, fam, G = pdpl.read_plink(\"data/EUR.22.test/1000G.EUR.22\")\n",
    "\n",
    "# This file contains the gene expression data we will be using\n",
    "gene_expr = pd.read_csv('data/GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# We also want to grab the gene annotations\n",
    "with gzip.open(\"data/gene_annot.txt.gz\", \"rt\") as f:\n",
    "    # Read the file content into a pandas DataFrame\n",
    "    gene_annotations = pd.read_csv(f, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0761f0fc-5b05-4bb2-a00c-790d09c42b11",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    }
   },
   "outputs": [],
   "source": [
    "# Combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df = pd.DataFrame(G.compute().transpose())\n",
    "geno_df.columns = bim['snp']  # Use SNP IDs as columns\n",
    "geno_df.index = fam['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t = gene_expr.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t.index.name = 'iid'\n",
    "matching_ids = gene_expr_t.index.intersection(fam['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = geno_df.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]\n",
    "\n",
    "# Split into train and test\n",
    "X_train, X_test, y_train, y_test = train_test_split(geno_df_filtered, gene_expr_filtered, test_size=0.25, random_state=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0855141f-d3ff-4736-a793-1df7c4d18a93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs138168, skipping.\n",
      "Insufficient model output for SNP rs138171, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs17001322, skipping.\n",
      "Insufficient model output for SNP rs5999980, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n",
      "Insufficient model output for SNP rs8142085, skipping.\n"
     ]
    }
   ],
   "source": [
    "# Grab genes from desired chromosomes\n",
    "chromosome = \"22\"\n",
    "chr22_genes = gene_expr[gene_expr['Chr'] == chromosome]['Gene_Symbol']\n",
    "\n",
    "# Initialize a list to store results for each SNP and gene as rows\n",
    "chr22_results_list = []\n",
    "\n",
    "for the_gene in chr22_genes:\n",
    "    gene_info = gene_expr[gene_expr['Gene_Symbol'] == the_gene]\n",
    "    snps_in_window = bim[(bim['chrom'] == chromosome) & \n",
    "                         (bim['pos'] >= int(gene_info['Coord'].iloc[0])-500000) & \n",
    "                         (bim['pos'] <= int(gene_info['Coord'].iloc[0])+500000)]\n",
    "    \n",
    "    # Get the SNP IDs from the filtered SNPs\n",
    "    snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "    # Filter the genotype DataFrame for the SNPs in the window\n",
    "    X_train_filtered = X_train.loc[:, snp_ids]\n",
    "\n",
    "    # Filter the gene expression data for the gene of interest\n",
    "    y_train_t = y_train.transpose()\n",
    "    y_train_filtered = y_train_t.loc[the_gene]\n",
    "\n",
    "    # Merge the gene expression data with genotype data\n",
    "    merged_df = pd.concat([X_train_filtered, y_train_filtered], axis=1)\n",
    "\n",
    "    # Rename the gene expression column for clarity\n",
    "    merged_df.rename(columns={the_gene: 'gene_expression'}, inplace=True)\n",
    "    \n",
    "    # Ensure gene expression values are numeric\n",
    "    y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "    for snp in X_train_filtered.columns:\n",
    "        X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "        X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "        # Drop rows with missing or invalid data (NaN)\n",
    "        valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "        # Run the regression only if valid data exists\n",
    "        if len(valid_data) > 0:\n",
    "            model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "\n",
    "            if len(model.pvalues) > 1:\n",
    "                # Append results for this SNP as a row in the results list\n",
    "                chr22_results_list.append({\n",
    "                    'Gene': the_gene,\n",
    "                    'SNP': snp,\n",
    "                    'p-value': model.pvalues.iloc[1],\n",
    "                    'beta': model.params.iloc[1],\n",
    "                    'r_squared': model.rsquared\n",
    "                })\n",
    "            else:\n",
    "                print(f\"Insufficient model output for SNP {snp}, skipping.\")\n",
    "\n",
    "# Convert the list of results into a DataFrame\n",
    "chr22_results_df = pd.DataFrame(chr22_results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "381ddce1-57bc-48e4-9f8b-b9853964e225",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr22_results_df.to_csv('chr22_eqtl_data.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb02810-1f9b-4389-8263-f8f7e1f8de29",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd6f830b-b79a-459a-8c26-75ccf4c69b1a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946353a5-02ad-40f8-8dca-a8b0d3873120",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24e5b5c8-826c-4d14-9a66-79fa170ea98f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6314c8c5-1e76-46f2-87ec-4ef765d5b60f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30cbe8f6-a85f-4185-bf99-fd9a87dcf3d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a2ad206-5b23-4365-8da8-a1591cd0dfde",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d32c106-4804-4bf7-bf58-fff31ac5a211",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the report doing all of the chromosomes:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a77c8caf-e6d7-4f7a-bd53-d141dd863093",
   "metadata": {},
   "source": [
    "### Chromosome 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "11790f8f-25b0-4046-a534-38984ba9e060",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:01<00:00,  2.14it/s]\n"
     ]
    }
   ],
   "source": [
    "bim1, fam1, G1 = pdpl.read_plink(\"data/EUR.1.test/1000G.EUR.1\")\n",
    "gene_expr1 = pd.read_csv('data/GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# Combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df1 = pd.DataFrame(G1.compute().transpose())\n",
    "geno_df1.columns = bim1['snp']  # Use SNP IDs as columns\n",
    "geno_df1.index = fam1['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t1 = gene_expr1.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t1.index.name = 'iid'\n",
    "matching_ids1 = gene_expr_t1.index.intersection(fam1['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered1 = geno_df1.loc[matching_ids1]\n",
    "gene_expr_filtered1 = gene_expr_t1.loc[matching_ids1]\n",
    "\n",
    "# Split into train and test\n",
    "X_train1, X_test1, y_train1, y_test1 = train_test_split(geno_df_filtered1, gene_expr_filtered1, test_size=0.25, random_state=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b14940bf-9af3-493b-9080-db3fba34aba9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs12075758, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs7531646, skipping.\n",
      "Insufficient model output for SNP rs12075758, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs7547835, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs7531646, skipping.\n",
      "Insufficient model output for SNP rs1484299, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs7531646, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs10912581, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs1475515, skipping.\n",
      "Insufficient model output for SNP rs6690221, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n",
      "Insufficient model output for SNP rs12075758, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs1484299, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n",
      "Insufficient model output for SNP rs7531646, skipping.\n",
      "Insufficient model output for SNP rs1484299, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs12075758, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs10912581, skipping.\n",
      "Insufficient model output for SNP rs7547835, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs2819390, skipping.\n",
      "Insufficient model output for SNP rs2819386, skipping.\n",
      "Insufficient model output for SNP rs10912581, skipping.\n",
      "Insufficient model output for SNP rs10912581, skipping.\n",
      "Insufficient model output for SNP rs12075758, skipping.\n",
      "Insufficient model output for SNP rs4259582, skipping.\n",
      "Insufficient model output for SNP rs2501420, skipping.\n",
      "Insufficient model output for SNP rs7537933, skipping.\n",
      "Insufficient model output for SNP rs17115864, skipping.\n",
      "Insufficient model output for SNP rs1484299, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs7531646, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs7531646, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs12065314, skipping.\n",
      "Insufficient model output for SNP rs11805314, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n",
      "Insufficient model output for SNP rs12075758, skipping.\n",
      "Insufficient model output for SNP rs2782831, skipping.\n",
      "Insufficient model output for SNP rs9429946, skipping.\n",
      "Insufficient model output for SNP rs16843437, skipping.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[66], line 35\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m snp \u001b[38;5;129;01min\u001b[39;00m X_train_filtered\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[1;32m     34\u001b[0m     X \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mto_numeric(merged_df[snp], errors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcoerce\u001b[39m\u001b[38;5;124m'\u001b[39m)  \u001b[38;5;66;03m# Convert SNP data to numeric\u001b[39;00m\n\u001b[0;32m---> 35\u001b[0m     X \u001b[38;5;241m=\u001b[39m \u001b[43msm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43madd_constant\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# Add an intercept for the linear regression\u001b[39;00m\n\u001b[1;32m     37\u001b[0m     \u001b[38;5;66;03m# Drop rows with missing or invalid data (NaN)\u001b[39;00m\n\u001b[1;32m     38\u001b[0m     valid_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([y, X], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\u001b[38;5;241m.\u001b[39mdropna()\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/statsmodels/tools/tools.py:185\u001b[0m, in \u001b[0;36madd_constant\u001b[0;34m(data, prepend, has_constant)\u001b[0m\n\u001b[1;32m    183\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m _is_using_pandas(data, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    184\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mstatsmodels\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtsa\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtsatools\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m add_trend\n\u001b[0;32m--> 185\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43madd_trend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtrend\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mc\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprepend\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprepend\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhas_constant\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhas_constant\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    187\u001b[0m \u001b[38;5;66;03m# Special case for NumPy\u001b[39;00m\n\u001b[1;32m    188\u001b[0m x \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(data)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/statsmodels/tsa/tsatools.py:162\u001b[0m, in \u001b[0;36madd_trend\u001b[0;34m(x, trend, prepend, has_constant)\u001b[0m\n\u001b[1;32m    160\u001b[0m     trendarr \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(trendarr, index\u001b[38;5;241m=\u001b[39mx\u001b[38;5;241m.\u001b[39mindex, columns\u001b[38;5;241m=\u001b[39mcolumns)\n\u001b[1;32m    161\u001b[0m     x \u001b[38;5;241m=\u001b[39m [trendarr, x]\n\u001b[0;32m--> 162\u001b[0m     x \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconcat\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m:\u001b[49m\u001b[43morder\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    164\u001b[0m     x \u001b[38;5;241m=\u001b[39m [trendarr, x]\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/reshape/concat.py:395\u001b[0m, in \u001b[0;36mconcat\u001b[0;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[1;32m    380\u001b[0m     copy \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    382\u001b[0m op \u001b[38;5;241m=\u001b[39m _Concatenator(\n\u001b[1;32m    383\u001b[0m     objs,\n\u001b[1;32m    384\u001b[0m     axis\u001b[38;5;241m=\u001b[39maxis,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    392\u001b[0m     sort\u001b[38;5;241m=\u001b[39msort,\n\u001b[1;32m    393\u001b[0m )\n\u001b[0;32m--> 395\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/reshape/concat.py:684\u001b[0m, in \u001b[0;36m_Concatenator.get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    680\u001b[0m             indexers[ax] \u001b[38;5;241m=\u001b[39m obj_labels\u001b[38;5;241m.\u001b[39mget_indexer(new_labels)\n\u001b[1;32m    682\u001b[0m     mgrs_indexers\u001b[38;5;241m.\u001b[39mappend((obj\u001b[38;5;241m.\u001b[39m_mgr, indexers))\n\u001b[0;32m--> 684\u001b[0m new_data \u001b[38;5;241m=\u001b[39m \u001b[43mconcatenate_managers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    685\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmgrs_indexers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnew_axes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconcat_axis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbm_axis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\n\u001b[1;32m    686\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    687\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcopy \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m using_copy_on_write():\n\u001b[1;32m    688\u001b[0m     new_data\u001b[38;5;241m.\u001b[39m_consolidate_inplace()\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/internals/concat.py:131\u001b[0m, in \u001b[0;36mconcatenate_managers\u001b[0;34m(mgrs_indexers, axes, concat_axis, copy)\u001b[0m\n\u001b[1;32m    124\u001b[0m \u001b[38;5;66;03m# Assertions disabled for performance\u001b[39;00m\n\u001b[1;32m    125\u001b[0m \u001b[38;5;66;03m# for tup in mgrs_indexers:\u001b[39;00m\n\u001b[1;32m    126\u001b[0m \u001b[38;5;66;03m#    # caller is responsible for ensuring this\u001b[39;00m\n\u001b[1;32m    127\u001b[0m \u001b[38;5;66;03m#    indexers = tup[1]\u001b[39;00m\n\u001b[1;32m    128\u001b[0m \u001b[38;5;66;03m#    assert concat_axis not in indexers\u001b[39;00m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m concat_axis \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 131\u001b[0m     mgrs \u001b[38;5;241m=\u001b[39m \u001b[43m_maybe_reindex_columns_na_proxy\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmgrs_indexers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneeds_copy\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    132\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m mgrs[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mconcat_horizontal(mgrs, axes)\n\u001b[1;32m    134\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mgrs_indexers) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m mgrs_indexers[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mnblocks \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/internals/concat.py:230\u001b[0m, in \u001b[0;36m_maybe_reindex_columns_na_proxy\u001b[0;34m(axes, mgrs_indexers, needs_copy)\u001b[0m\n\u001b[1;32m    220\u001b[0m         mgr \u001b[38;5;241m=\u001b[39m mgr\u001b[38;5;241m.\u001b[39mreindex_indexer(\n\u001b[1;32m    221\u001b[0m             axes[i],\n\u001b[1;32m    222\u001b[0m             indexers[i],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    227\u001b[0m             use_na_proxy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,  \u001b[38;5;66;03m# only relevant for i==0\u001b[39;00m\n\u001b[1;32m    228\u001b[0m         )\n\u001b[1;32m    229\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m needs_copy \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m indexers:\n\u001b[0;32m--> 230\u001b[0m         mgr \u001b[38;5;241m=\u001b[39m \u001b[43mmgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    232\u001b[0m     new_mgrs\u001b[38;5;241m.\u001b[39mappend(mgr)\n\u001b[1;32m    233\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m new_mgrs\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/internals/managers.py:586\u001b[0m, in \u001b[0;36mBaseBlockManager.copy\u001b[0;34m(self, deep)\u001b[0m\n\u001b[1;32m    583\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcopy_func\u001b[39m(ax):\n\u001b[1;32m    584\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m ax\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;28;01mif\u001b[39;00m deep \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m ax\u001b[38;5;241m.\u001b[39mview()\n\u001b[0;32m--> 586\u001b[0m     new_axes \u001b[38;5;241m=\u001b[39m \u001b[43m[\u001b[49m\u001b[43mcopy_func\u001b[49m\u001b[43m(\u001b[49m\u001b[43max\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43max\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maxes\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    587\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    588\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m using_copy_on_write():\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/internals/managers.py:586\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    583\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcopy_func\u001b[39m(ax):\n\u001b[1;32m    584\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m ax\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;28;01mif\u001b[39;00m deep \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m ax\u001b[38;5;241m.\u001b[39mview()\n\u001b[0;32m--> 586\u001b[0m     new_axes \u001b[38;5;241m=\u001b[39m [\u001b[43mcopy_func\u001b[49m\u001b[43m(\u001b[49m\u001b[43max\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m ax \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes]\n\u001b[1;32m    587\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    588\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m using_copy_on_write():\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/internals/managers.py:584\u001b[0m, in \u001b[0;36mBaseBlockManager.copy.<locals>.copy_func\u001b[0;34m(ax)\u001b[0m\n\u001b[1;32m    583\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcopy_func\u001b[39m(ax):\n\u001b[0;32m--> 584\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ax\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;28;01mif\u001b[39;00m deep \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m \u001b[43max\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mview\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/indexes/base.py:1040\u001b[0m, in \u001b[0;36mIndex.view\u001b[0;34m(self, cls)\u001b[0m\n\u001b[1;32m   1030\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1031\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1032\u001b[0m             \u001b[38;5;66;03m# GH#55709\u001b[39;00m\n\u001b[1;32m   1033\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPassing a type in \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.view is deprecated \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1037\u001b[0m             stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m   1038\u001b[0m         )\n\u001b[0;32m-> 1040\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_view\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1041\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(result, Index):\n\u001b[1;32m   1042\u001b[0m     result\u001b[38;5;241m.\u001b[39m_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_id\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/indexes/base.py:777\u001b[0m, in \u001b[0;36mIndex._view\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    773\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_view\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Self:\n\u001b[1;32m    774\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    775\u001b[0m \u001b[38;5;124;03m    fastpath to make a shallow copy, i.e. new object with same data.\u001b[39;00m\n\u001b[1;32m    776\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 777\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_simple_new\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_values\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrefs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_references\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    779\u001b[0m     result\u001b[38;5;241m.\u001b[39m_cache \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cache\n\u001b[1;32m    780\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/core/indexes/base.py:670\u001b[0m, in \u001b[0;36mIndex._simple_new\u001b[0;34m(cls, values, name, refs)\u001b[0m\n\u001b[1;32m    668\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    669\u001b[0m     result\u001b[38;5;241m.\u001b[39m_references \u001b[38;5;241m=\u001b[39m BlockValuesRefs()\n\u001b[0;32m--> 670\u001b[0m \u001b[43mresult\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_references\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43madd_index_reference\u001b[49m\u001b[43m(\u001b[49m\u001b[43mresult\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Grab genes from desired chromosomes\n",
    "chromosome = \"1\"\n",
    "chr1_genes = gene_expr1[gene_expr1['Chr'] == chromosome]['Gene_Symbol']\n",
    "\n",
    "# Initialize a list to store results for each SNP and gene as rows\n",
    "chr1_results_list = []\n",
    "\n",
    "for the_gene in chr1_genes:\n",
    "    gene_info = gene_expr1[gene_expr1['Gene_Symbol'] == the_gene]\n",
    "    snps_in_window = bim1[(bim1['chrom'] == chromosome) & \n",
    "                         (bim1['pos'] >= int(gene_info['Coord'].iloc[0])-500000) & \n",
    "                         (bim1['pos'] <= int(gene_info['Coord'].iloc[0])+500000)]\n",
    "    \n",
    "    # Get the SNP IDs from the filtered SNPs\n",
    "    snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "    # Filter the genotype DataFrame for the SNPs in the window\n",
    "    X_train_filtered = X_train1.loc[:, snp_ids]\n",
    "\n",
    "    # Filter the gene expression data for the gene of interest\n",
    "    y_train_t = y_train1.transpose()\n",
    "    y_train_filtered = y_train_t.loc[the_gene]\n",
    "\n",
    "    # Merge the gene expression data with genotype data\n",
    "    merged_df = pd.concat([X_train_filtered, y_train_filtered], axis=1)\n",
    "\n",
    "    # Rename the gene expression column for clarity\n",
    "    merged_df.rename(columns={the_gene: 'gene_expression'}, inplace=True)\n",
    "    \n",
    "    # Ensure gene expression values are numeric\n",
    "    y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "    for snp in X_train_filtered.columns:\n",
    "        X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "        X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "        # Drop rows with missing or invalid data (NaN)\n",
    "        valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "        # Run the regression only if valid data exists\n",
    "        if len(valid_data) > 0:\n",
    "            model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "\n",
    "            if len(model.pvalues) > 1:\n",
    "                # Append results for this SNP as a row in the results list\n",
    "                chr1_results_list.append({\n",
    "                    'Gene': the_gene,\n",
    "                    'SNP': snp,\n",
    "                    'p-value': model.pvalues.iloc[1],\n",
    "                    'beta': model.params.iloc[1],\n",
    "                    'r_squared': model.rsquared\n",
    "                })\n",
    "            else:\n",
    "                print(f\"Insufficient model output for SNP {snp}, skipping.\")\n",
    "\n",
    "# Convert the list of results into a DataFrame\n",
    "chr1_results_df = pd.DataFrame(chr1_results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f4d3e7c-5c26-4e6a-a631-f6b1bae87fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr1_results_df.to_csv('chr1_eqtl_data.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21ff320b-cf23-4aa5-8a4b-2f342dcdbf8f",
   "metadata": {},
   "source": [
    "### Chromosome 14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42909491-8c16-4820-90a5-f7bb6ca46729",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim14, fam14, G14 = pdpl.read_plink(\"data/EUR.14.test/1000G.EUR.14\")\n",
    "gene_expr14 = pd.read_csv('data/GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# Combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df14 = pd.DataFrame(G1.compute().transpose())\n",
    "geno_df14.columns = bim14['snp']  # Use SNP IDs as columns\n",
    "geno_df14.index = fam14['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t14 = gene_expr14.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t14.index.name = 'iid'\n",
    "matching_ids14 = gene_expr_t14.index.intersection(fam1['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered14 = geno_df14.loc[matching_ids1]\n",
    "gene_expr_filtered14 = gene_expr_t14.loc[matching_ids1]\n",
    "\n",
    "# Split into train and test\n",
    "X_train14, X_test14, y_train14, y_test14 = train_test_split(geno_df_filtered14, gene_expr_filtered14, test_size=0.25, random_state=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67e95d77-e438-4e81-8715-4f40d9a79e99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Grab genes from desired chromosomes\n",
    "chromosome = \"14\"\n",
    "chr14_genes = gene_expr1[gene_expr1['Chr'] == chromosome]['Gene_Symbol']\n",
    "\n",
    "# Initialize a list to store results for each SNP and gene as rows\n",
    "chr14_results_list = []\n",
    "\n",
    "for the_gene in chr14_genes:\n",
    "    gene_info = gene_expr1[gene_expr14['Gene_Symbol'] == the_gene]\n",
    "    snps_in_window = bim14[(bim14['chrom'] == chromosome) & \n",
    "                         (bim14['pos'] >= int(gene_info['Coord'].iloc[0])-500000) & \n",
    "                         (bim14['pos'] <= int(gene_info['Coord'].iloc[0])+500000)]\n",
    "    \n",
    "    # Get the SNP IDs from the filtered SNPs\n",
    "    snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "    # Filter the genotype DataFrame for the SNPs in the window\n",
    "    X_train_filtered = X_train14.loc[:, snp_ids]\n",
    "\n",
    "    # Filter the gene expression data for the gene of interest\n",
    "    y_train_t = y_train14.transpose()\n",
    "    y_train_filtered = y_train_t.loc[the_gene]\n",
    "\n",
    "    # Merge the gene expression data with genotype data\n",
    "    merged_df = pd.concat([X_train_filtered, y_train_filtered], axis=1)\n",
    "\n",
    "    # Rename the gene expression column for clarity\n",
    "    merged_df.rename(columns={the_gene: 'gene_expression'}, inplace=True)\n",
    "    \n",
    "    # Ensure gene expression values are numeric\n",
    "    y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "    for snp in X_train_filtered.columns:\n",
    "        X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "        X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "        # Drop rows with missing or invalid data (NaN)\n",
    "        valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "        # Run the regression only if valid data exists\n",
    "        if len(valid_data) > 0:\n",
    "            model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "\n",
    "            if len(model.pvalues) > 1:\n",
    "                # Append results for this SNP as a row in the results list\n",
    "                chr14_results_list.append({\n",
    "                    'Gene': the_gene,\n",
    "                    'SNP': snp,\n",
    "                    'p-value': model.pvalues.iloc[1],\n",
    "                    'beta': model.params.iloc[1],\n",
    "                    'r_squared': model.rsquared\n",
    "                })\n",
    "            else:\n",
    "                print(f\"Insufficient model output for SNP {snp}, skipping.\")\n",
    "\n",
    "# Convert the list of results into a DataFrame\n",
    "chr14_results_df = pd.DataFrame(chr14_results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa288455-0650-4958-ba3b-b508edae3912",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr14_results_df.to_csv('chr14_eqtl_data.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e6a3bc-97af-4fa6-8d1b-9ca2d8e0205f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
