{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819dc8dc",
   "metadata": {},
   "source": [
    "## Start with Imports and Loading in the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "5e12fb97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: plink in /home/jhauk/.local/lib/python3.9/site-packages (2.4.2)\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import pandas_plink as pdpl\n",
    "import statsmodels.api as sm\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "import subprocess\n",
    "from sklearn.metrics import r2_score\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "!pip install plink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "37b55f25",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files:   0%|          | 0/3 [00:00<?, ?it/s]/home/jhauk/.local/lib/python3.9/site-packages/pandas_plink/_read.py:338: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = read_csv(\n",
      "Mapping files:  33%|███▎      | 1/3 [00:00<00:00,  8.64it/s]/home/jhauk/.local/lib/python3.9/site-packages/pandas_plink/_read.py:338: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = read_csv(\n",
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 17.27it/s]\n"
     ]
    }
   ],
   "source": [
    "# There are three file types for genetic data that interest us\n",
    "# Multiple different samples of these files can be found on 1000 Genomes\n",
    "# In the interest of size and reproducibility we will use just EUR.1 in the beginning\n",
    "bim, fam, G = pdpl.read_plink(\"EUR.1.test/1000G.EUR.1\")\n",
    "\n",
    "# This file contains the gene expression data we will be using\n",
    "gene_expr = pd.read_csv('GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# We also want to grab the gene annotations\n",
    "with gzip.open(\"gene_annot.txt.gz\", \"rt\") as f:\n",
    "    # Read the file content into a pandas DataFrame\n",
    "    gene_annotations = pd.read_csv(f, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52cf419a",
   "metadata": {},
   "source": [
    "## Data Wrangling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e2b19c19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want to combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df = pd.DataFrame(G.compute().transpose())\n",
    "geno_df.columns = bim['snp']  # Use SNP IDs as columns\n",
    "geno_df.index = fam['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t = gene_expr.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t.index.name = 'iid'\n",
    "matching_ids = gene_expr_t.index.intersection(fam['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = geno_df.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e532415",
   "metadata": {},
   "source": [
    "## Test eQTL Analysis for a Single Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2db94f91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most Significant Results:\n",
      "            p-value      beta  r_squared\n",
      "rs4970420  0.004872 -0.016493   0.022943\n",
      "rs9442373  0.030045  0.010051   0.013688\n",
      "rs3748594  0.038706 -0.023893   0.012437\n",
      "rs3748593  0.052245 -0.022778   0.010973\n",
      "rs6696281  0.071271 -0.017948   0.009482\n",
      "\n",
      "All Results:\n",
      "             p-value      beta  r_squared\n",
      "rs3094315   0.179067 -0.008622   0.005273\n",
      "rs3131972   0.150053 -0.009219   0.006048\n",
      "rs3131969   0.365582 -0.006137   0.002394\n",
      "rs1048488   0.208763 -0.007981   0.004616\n",
      "rs3115850   0.223496 -0.007717   0.004330\n",
      "...              ...       ...        ...\n",
      "rs1536168   0.104462 -0.018155   0.007689\n",
      "rs3737717   0.942082 -0.001428   0.000015\n",
      "rs12103     0.541356 -0.003543   0.001092\n",
      "rs12142199  0.302081 -0.005895   0.003114\n",
      "rs34054191  0.759153 -0.003741   0.000275\n",
      "\n",
      "[161 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# Randomly select a gene to test:\n",
    "gene_of_interest = 'ENSG00000152931.6'\n",
    "\n",
    "# Adjust window according to gene location\n",
    "# [We will automate this process in the next step]\n",
    "snps_in_window = bim[(bim['chrom'] == \"1\") & \n",
    "                         (bim['pos'] >= 252566) & \n",
    "                         (bim['pos'] <= 1252566)]\n",
    "\n",
    "# Get the SNP IDs from the filtered SNPs\n",
    "snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "# Filter the genotype DataFrame for the SNPs in the window\n",
    "genotype_data_window = geno_df.loc[:, snp_ids]\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = genotype_data_window.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]\n",
    "\n",
    "# Filter the gene expression data for the gene of interest\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids, gene_of_interest]\n",
    "\n",
    "# Merge the gene expression data with genotype data\n",
    "merged_df = pd.concat([gene_expr_filtered, geno_df_filtered], axis=1)\n",
    "\n",
    "# Rename the gene expression column for clarity\n",
    "merged_df.rename(columns={gene_of_interest: 'gene_expression'}, inplace=True)\n",
    "\n",
    "# Ensure gene expression values are numeric\n",
    "y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "# Iterate over each SNP column in the genotype data (independent variables)\n",
    "results = {}\n",
    "for snp in geno_df_filtered.columns:\n",
    "    X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "    X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "    # Drop rows with missing or invalid data (NaN)\n",
    "    valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "    # Run the regression only if valid data exists\n",
    "    if len(valid_data) > 0:\n",
    "        model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "        \n",
    "        # Store the result for this SNP\n",
    "        results[snp] = {\n",
    "            'p-value': model.pvalues.iloc[1],  # Use .iloc to access the p-value for the SNP\n",
    "            'beta': model.params.iloc[1],  # Use .iloc to access the beta coefficient for the SNP\n",
    "            'r_squared': model.rsquared  # R-squared value\n",
    "        }\n",
    "\n",
    "# Convert results to a DataFrame for easier analysis\n",
    "results_df = pd.DataFrame(results).transpose()\n",
    "\n",
    "# Sort by p-value to find significant eQTLs\n",
    "significant_eqtls = results_df.sort_values('p-value').head()\n",
    "print(\"Most Significant Results:\")\n",
    "print(significant_eqtls)\n",
    "print(\"\\nAll Results:\")\n",
    "print(results_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcfe9113",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust the results for plotting purposes\n",
    "results_df['neg_log10_p'] = -np.log10(results_df['p-value'])\n",
    "results_df1 = pd.merge(results_df, snps_in_window, left_index=True, right_on='snp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17e6eb5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a locus zoom plot to look at results\n",
    "# Set the aesthetics for the plot\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "# Create the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(data=results_df1, x='pos', y='neg_log10_p', color='blue', s=100)\n",
    "\n",
    "# Adding labels and title\n",
    "plt.title('LocusZoom Plot of eQTL Analysis', fontsize=16)\n",
    "plt.xlabel('Genomic Coordinate (bp)', fontsize=14)\n",
    "plt.ylabel('-log10(p-value)', fontsize=14)\n",
    "\n",
    "# Adding a horizontal line for significance threshold (e.g., p < 0.05)\n",
    "significance_threshold = -np.log10(0.05)\n",
    "plt.axhline(y=significance_threshold, color='red', linestyle='--', label='p < 0.05')\n",
    "\n",
    "# Customize x-axis limits (optional)\n",
    "plt.xlim([results_df1['pos'].min() - 50000, results_df1['pos'].max() + 50000])\n",
    "\n",
    "# Add legend\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4687b955",
   "metadata": {},
   "source": [
    "## Perform cis-eQTL for All the Genes on a Chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "4f361de0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Mapping files:   0%|          | 0/3 [00:00<?, ?it/s]\u001b[A/home/jhauk/.local/lib/python3.9/site-packages/pandas_plink/_read.py:338: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = read_csv(\n",
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 48.29it/s]\n"
     ]
    }
   ],
   "source": [
    "# We will now repeat some of the process from above to transition to using chromosome 14\n",
    "# This will help us in terms of size and time\n",
    "bim, fam, G = pdpl.read_plink(\"EUR.14.test/1000G.EUR.14\")\n",
    "\n",
    "# This file contains the gene expression data we will be using\n",
    "gene_expr = pd.read_csv('GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# We also want to grab the gene annotations\n",
    "with gzip.open(\"gene_annot.txt.gz\", \"rt\") as f:\n",
    "    # Read the file content into a pandas DataFrame\n",
    "    gene_annotations = pd.read_csv(f, sep=\"\\t\")\n",
    "    \n",
    "# We want to combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df = pd.DataFrame(G.compute().transpose())\n",
    "geno_df.columns = bim['snp']  # Use SNP IDs as columns\n",
    "geno_df.index = fam['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t = gene_expr.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t.index.name = 'iid'\n",
    "matching_ids = gene_expr_t.index.intersection(fam['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = geno_df.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9f38260d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000243417.1 completed\n",
      "ENSG00000152969.11 completed\n",
      "ENSG00000249212.1 completed\n",
      "ENSG00000251297.1 completed\n",
      "ENSG00000248155.1 completed\n",
      "ENSG00000178950.11 completed\n",
      "ENSG00000145365.8 completed\n",
      "ENSG00000164096.10 completed\n",
      "ENSG00000186222.3 completed\n",
      "ENSG00000248049.2 completed\n",
      "ENSG00000164167.5 completed\n",
      "ENSG00000168785.3 completed\n",
      "ENSG00000225892.3 completed\n",
      "ENSG00000249952.1 completed\n",
      "ENSG00000197410.6 completed\n",
      "ENSG00000251555.1 completed\n",
      "ENSG00000248697.1 completed\n",
      "ENSG00000152785.5 completed\n",
      "ENSG00000151470.7 completed\n",
      "ENSG00000241281.1 completed\n",
      "ENSG00000185049.7 completed\n",
      "ENSG00000250910.1 completed\n",
      "ENSG00000223305.1 completed\n",
      "ENSG00000109163.6 completed\n",
      "ENSG00000213522.3 completed\n",
      "ENSG00000151718.11 completed\n",
      "ENSG00000061918.8 completed\n",
      "ENSG00000186212.2 completed\n",
      "ENSG00000182903.11 completed\n",
      "ENSG00000251587.1 completed\n",
      "ENSG00000157796.12 completed\n",
      "ENSG00000236530.1 completed\n",
      "ENSG00000087008.10 completed\n",
      "ENSG00000248394.1 completed\n",
      "ENSG00000213492.2 completed\n",
      "ENSG00000163281.6 completed\n",
      "ENSG00000248632.1 completed\n",
      "ENSG00000164056.6 completed\n",
      "ENSG00000250666.1 completed\n",
      "ENSG00000109270.8 completed\n",
      "ENSG00000169245.4 completed\n",
      "ENSG00000250312.2 completed\n",
      "ENSG00000249348.1 completed\n",
      "ENSG00000244004.1 completed\n",
      "ENSG00000151729.5 completed\n",
      "ENSG00000249502.1 completed\n",
      "ENSG00000249241.1 completed\n",
      "ENSG00000250030.1 completed\n",
      "ENSG00000164105.3 completed\n",
      "ENSG00000174227.10 completed\n",
      "ENSG00000132406.7 completed\n",
      "ENSG00000173011.10 completed\n",
      "ENSG00000172955.13 completed\n",
      "ENSG00000109794.9 completed\n",
      "ENSG00000138777.15 completed\n",
      "ENSG00000251596.1 completed\n",
      "ENSG00000239983.1 completed\n",
      "ENSG00000168769.7 completed\n",
      "ENSG00000224207.2 completed\n",
      "ENSG00000083857.8 completed\n",
      "ENSG00000249863.1 completed\n",
      "ENSG00000109670.8 completed\n",
      "ENSG00000251332.1 completed\n",
      "ENSG00000250656.1 completed\n",
      "ENSG00000251571.1 completed\n",
      "ENSG00000205595.3 completed\n",
      "ENSG00000137601.11 completed\n",
      "ENSG00000109180.10 completed\n",
      "ENSG00000260296.1 completed\n",
      "ENSG00000236699.3 completed\n",
      "ENSG00000168924.9 completed\n",
      "ENSG00000250626.2 completed\n",
      "ENSG00000075539.9 completed\n",
      "ENSG00000185477.3 completed\n",
      "ENSG00000248645.1 completed\n",
      "ENSG00000197057.4 completed\n",
      "ENSG00000181877.3 completed\n",
      "ENSG00000188848.10 completed\n",
      "ENSG00000164161.5 completed\n",
      "ENSG00000186352.4 completed\n",
      "ENSG00000250220.1 completed\n",
      "ENSG00000224807.5 completed\n",
      "ENSG00000109819.4 completed\n",
      "ENSG00000138778.6 completed\n",
      "ENSG00000172007.5 completed\n",
      "ENSG00000242640.1 completed\n",
      "ENSG00000121895.7 completed\n",
      "ENSG00000263327.1 completed\n",
      "ENSG00000250471.1 completed\n",
      "ENSG00000164070.6 completed\n",
      "ENSG00000164134.6 completed\n",
      "ENSG00000164111.9 completed\n",
      "ENSG00000164323.8 completed\n",
      "ENSG00000251230.1 completed\n",
      "ENSG00000198498.5 completed\n",
      "ENSG00000174720.11 completed\n",
      "ENSG00000251463.1 completed\n",
      "ENSG00000245562.2 completed\n",
      "ENSG00000013288.4 completed\n",
      "ENSG00000205794.3 completed\n",
      "ENSG00000145337.4 completed\n",
      "ENSG00000109534.12 completed\n",
      "ENSG00000109618.7 completed\n",
      "ENSG00000260526.1 completed\n",
      "ENSG00000174796.8 completed\n",
      "ENSG00000145244.7 completed\n",
      "ENSG00000247950.2 completed\n",
      "ENSG00000170871.7 completed\n",
      "ENSG00000164342.6 completed\n",
      "ENSG00000163069.8 completed\n",
      "ENSG00000138821.8 completed\n",
      "ENSG00000260091.1 completed\n",
      "ENSG00000178988.10 completed\n",
      "ENSG00000137449.10 completed\n",
      "ENSG00000248113.1 completed\n",
      "ENSG00000053900.6 completed\n",
      "ENSG00000247708.2 completed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files:   0%|          | 0/3 [01:46<?, ?it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000227725.2 completed\n",
      "ENSG00000138722.5 completed\n",
      "ENSG00000074966.6 completed\n",
      "ENSG00000249934.1 completed\n",
      "ENSG00000169020.5 completed\n",
      "ENSG00000181359.5 completed\n",
      "ENSG00000214980.4 completed\n",
      "ENSG00000163257.6 completed\n",
      "ENSG00000156234.7 completed\n",
      "ENSG00000251432.2 completed\n",
      "ENSG00000249844.1 completed\n",
      "ENSG00000109775.6 completed\n",
      "ENSG00000251259.1 completed\n",
      "ENSG00000249244.1 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000072201.8 completed\n",
      "ENSG00000178458.4 completed\n",
      "ENSG00000163743.9 completed\n",
      "ENSG00000152795.12 completed\n",
      "ENSG00000174607.6 completed\n",
      "ENSG00000242358.1 completed\n",
      "ENSG00000213434.2 completed\n",
      "ENSG00000250131.1 completed\n",
      "ENSG00000132405.14 completed\n",
      "ENSG00000109436.7 completed\n",
      "ENSG00000183432.6 completed\n",
      "ENSG00000232327.3 completed\n",
      "ENSG00000174125.3 completed\n",
      "ENSG00000250413.1 completed\n",
      "ENSG00000169989.2 completed\n",
      "ENSG00000145349.12 completed\n",
      "ENSG00000250980.1 completed\n",
      "ENSG00000258013.1 completed\n",
      "ENSG00000164031.12 completed\n",
      "ENSG00000085871.4 completed\n",
      "ENSG00000013810.13 completed\n",
      "ENSG00000250327.1 completed\n",
      "ENSG00000181982.11 completed\n",
      "ENSG00000163132.6 completed\n",
      "ENSG00000171234.9 completed\n",
      "ENSG00000180712.3 completed\n",
      "ENSG00000153147.4 completed\n",
      "ENSG00000249896.1 completed\n",
      "ENSG00000250566.1 completed\n",
      "ENSG00000138709.13 completed\n",
      "ENSG00000235043.2 completed\n",
      "ENSG00000121210.11 completed\n",
      "ENSG00000189184.7 completed\n",
      "ENSG00000163625.11 completed\n",
      "ENSG00000152784.9 completed\n",
      "ENSG00000052802.8 completed\n",
      "ENSG00000163319.6 completed\n",
      "ENSG00000249239.1 completed\n",
      "ENSG00000145241.6 completed\n",
      "ENSG00000251308.1 completed\n",
      "ENSG00000241877.1 completed\n",
      "ENSG00000225455.3 completed\n",
      "ENSG00000244146.1 completed\n",
      "ENSG00000005059.11 completed\n",
      "ENSG00000249207.1 completed\n",
      "ENSG00000109381.13 completed\n",
      "ENSG00000004468.6 completed\n",
      "ENSG00000047662.4 completed\n",
      "ENSG00000188981.6 completed\n",
      "ENSG00000250033.1 completed\n",
      "ENSG00000145476.10 completed\n",
      "ENSG00000109184.10 completed\n",
      "ENSG00000168556.5 completed\n",
      "ENSG00000248213.1 completed\n",
      "ENSG00000109790.11 completed\n",
      "ENSG00000214846.4 completed\n",
      "ENSG00000169299.8 completed\n",
      "ENSG00000164124.6 completed\n",
      "ENSG00000135220.6 completed\n",
      "ENSG00000168228.9 completed\n",
      "ENSG00000155275.14 completed\n",
      "ENSG00000109332.15 completed\n",
      "ENSG00000084093.10 completed\n",
      "ENSG00000170088.7 completed\n",
      "ENSG00000151612.10 completed\n",
      "ENSG00000261490.1 completed\n",
      "ENSG00000128059.4 completed\n",
      "ENSG00000241651.3 completed\n",
      "ENSG00000249920.1 completed\n",
      "ENSG00000138814.12 completed\n",
      "ENSG00000251284.1 completed\n",
      "ENSG00000213851.3 completed\n",
      "ENSG00000248401.1 completed\n",
      "ENSG00000214062.5 completed\n",
      "ENSG00000171566.7 completed\n",
      "ENSG00000249679.1 completed\n",
      "ENSG00000151466.7 completed\n",
      "ENSG00000226950.2 completed\n",
      "ENSG00000159733.9 completed\n",
      "ENSG00000083896.7 completed\n",
      "ENSG00000138686.5 completed\n",
      "ENSG00000091490.6 completed\n",
      "ENSG00000163697.11 completed\n",
      "ENSG00000249356.1 completed\n",
      "ENSG00000244538.1 completed\n",
      "ENSG00000245748.1 completed\n",
      "ENSG00000059691.7 completed\n",
      "ENSG00000245213.2 completed\n",
      "ENSG00000172403.6 completed\n",
      "ENSG00000164306.6 completed\n",
      "ENSG00000250325.1 completed\n",
      "ENSG00000164163.6 completed\n",
      "ENSG00000002549.7 completed\n",
      "ENSG00000173013.4 completed\n",
      "ENSG00000242169.1 completed\n",
      "ENSG00000240669.1 completed\n",
      "ENSG00000168843.9 completed\n",
      "ENSG00000174792.6 completed\n",
      "ENSG00000218336.3 completed\n",
      "ENSG00000227118.3 completed\n",
      "ENSG00000249509.1 completed\n",
      "ENSG00000052795.8 completed\n",
      "ENSG00000205940.5 completed\n",
      "ENSG00000014824.7 completed\n",
      "ENSG00000153064.7 completed\n",
      "ENSG00000138640.10 completed\n",
      "ENSG00000231160.3 completed\n",
      "ENSG00000145416.9 completed\n",
      "ENSG00000168743.8 completed\n",
      "ENSG00000109466.9 completed\n",
      "ENSG00000243199.1 completed\n",
      "ENSG00000245958.2 completed\n",
      "ENSG00000138785.10 completed\n",
      "ENSG00000251313.1 completed\n",
      "ENSG00000151247.8 completed\n",
      "ENSG00000145293.9 completed\n",
      "ENSG00000109743.6 completed\n",
      "ENSG00000251669.1 completed\n",
      "ENSG00000163633.6 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000134853.7 completed\n",
      "ENSG00000109536.6 completed\n",
      "ENSG00000251429.1 completed\n",
      "ENSG00000138794.5 completed\n",
      "ENSG00000150961.10 completed\n",
      "ENSG00000138688.9 completed\n",
      "ENSG00000145439.7 completed\n",
      "ENSG00000248122.1 completed\n",
      "ENSG00000226439.3 completed\n",
      "ENSG00000250896.1 completed\n",
      "ENSG00000250317.3 completed\n",
      "ENSG00000127419.10 completed\n",
      "ENSG00000154556.13 completed\n",
      "ENSG00000168818.5 completed\n",
      "ENSG00000249138.1 completed\n",
      "ENSG00000248336.1 completed\n",
      "ENSG00000179979.6 completed\n",
      "ENSG00000123737.8 completed\n",
      "ENSG00000164142.11 completed\n",
      "ENSG00000248571.1 completed\n",
      "ENSG00000185818.7 completed\n",
      "ENSG00000249548.1 completed\n",
      "ENSG00000248988.1 completed\n",
      "ENSG00000173085.9 completed\n",
      "ENSG00000251595.3 completed\n",
      "ENSG00000145386.5 completed\n",
      "ENSG00000145248.6 completed\n",
      "ENSG00000234841.2 completed\n",
      "ENSG00000213406.3 completed\n",
      "ENSG00000164114.13 completed\n",
      "ENSG00000129128.8 completed\n",
      "ENSG00000121089.4 completed\n",
      "ENSG00000249056.1 completed\n",
      "ENSG00000215375.2 completed\n",
      "ENSG00000163626.12 completed\n",
      "ENSG00000250768.1 completed\n",
      "ENSG00000145391.7 completed\n",
      "ENSG00000172493.15 completed\n",
      "ENSG00000234386.3 completed\n",
      "ENSG00000248866.1 completed\n",
      "ENSG00000181381.9 completed\n",
      "ENSG00000250412.1 completed\n",
      "ENSG00000178403.3 completed\n",
      "ENSG00000248774.1 completed\n",
      "ENSG00000245685.2 completed\n",
      "ENSG00000196620.4 completed\n",
      "ENSG00000047365.6 completed\n",
      "ENSG00000169019.9 completed\n",
      "ENSG00000254360.1 completed\n",
      "ENSG00000229204.2 completed\n",
      "ENSG00000243280.2 completed\n",
      "ENSG00000138735.11 completed\n",
      "ENSG00000109771.11 completed\n",
      "ENSG00000249532.1 completed\n",
      "ENSG00000184985.12 completed\n",
      "ENSG00000090989.13 completed\n",
      "ENSG00000151790.4 completed\n",
      "ENSG00000164116.12 completed\n",
      "ENSG00000159788.13 completed\n",
      "ENSG00000248546.3 completed\n",
      "ENSG00000151012.9 completed\n",
      "ENSG00000213370.3 completed\n",
      "ENSG00000145358.2 completed\n",
      "ENSG00000250610.1 completed\n",
      "ENSG00000171503.7 completed\n",
      "ENSG00000087266.11 completed\n",
      "ENSG00000250950.1 completed\n",
      "ENSG00000138750.10 completed\n",
      "ENSG00000109519.7 completed\n",
      "ENSG00000229450.2 completed\n",
      "ENSG00000234125.3 completed\n",
      "ENSG00000164169.8 completed\n",
      "ENSG00000145284.7 completed\n",
      "ENSG00000228981.3 completed\n",
      "ENSG00000248890.1 completed\n",
      "ENSG00000249700.1 completed\n",
      "ENSG00000109654.10 completed\n",
      "ENSG00000063978.10 completed\n",
      "ENSG00000145348.11 completed\n",
      "ENSG00000128045.5 completed\n",
      "ENSG00000157869.10 completed\n",
      "ENSG00000153130.12 completed\n",
      "ENSG00000168152.7 completed\n",
      "ENSG00000077684.10 completed\n",
      "ENSG00000189308.6 completed\n",
      "ENSG00000248196.1 completed\n",
      "ENSG00000145354.5 completed\n",
      "ENSG00000164099.3 completed\n",
      "ENSG00000196782.7 completed\n",
      "ENSG00000109736.10 completed\n",
      "ENSG00000251288.1 completed\n",
      "ENSG00000109606.7 completed\n",
      "ENSG00000250215.1 completed\n",
      "ENSG00000248785.1 completed\n",
      "ENSG00000248180.1 completed\n",
      "ENSG00000145331.9 completed\n",
      "ENSG00000138685.7 completed\n",
      "ENSG00000164023.10 completed\n",
      "ENSG00000261643.1 completed\n",
      "ENSG00000244459.2 completed\n",
      "ENSG00000159692.11 completed\n",
      "ENSG00000248778.1 completed\n",
      "ENSG00000251455.1 completed\n",
      "ENSG00000109321.6 completed\n",
      "ENSG00000241741.1 completed\n",
      "ENSG00000230069.3 completed\n",
      "ENSG00000138678.6 completed\n",
      "ENSG00000151611.8 completed\n",
      "ENSG00000170509.7 completed\n",
      "ENSG00000138698.9 completed\n",
      "ENSG00000251333.2 completed\n",
      "ENSG00000132464.7 completed\n",
      "ENSG00000205100.2 completed\n",
      "ENSG00000248641.1 completed\n",
      "ENSG00000179967.10 completed\n",
      "ENSG00000250185.1 completed\n",
      "ENSG00000151623.9 completed\n",
      "ENSG00000250568.1 completed\n",
      "ENSG00000185619.13 completed\n",
      "ENSG00000213448.3 completed\n",
      "ENSG00000178972.3 completed\n",
      "ENSG00000248590.2 completed\n",
      "ENSG00000249823.1 completed\n",
      "ENSG00000253917.2 completed\n",
      "ENSG00000168936.6 completed\n",
      "ENSG00000250410.1 completed\n",
      "ENSG00000048342.10 completed\n",
      "ENSG00000145425.5 completed\n",
      "ENSG00000182552.10 completed\n",
      "ENSG00000249334.1 completed\n",
      "ENSG00000260519.1 completed\n",
      "ENSG00000118564.10 completed\n",
      "ENSG00000248115.1 completed\n",
      "ENSG00000248327.1 completed\n",
      "ENSG00000135605.7 completed\n",
      "ENSG00000205208.4 completed\n",
      "ENSG00000256143.1 completed\n",
      "ENSG00000250321.1 completed\n",
      "ENSG00000145332.8 completed\n",
      "ENSG00000164117.7 completed\n",
      "ENSG00000249347.1 completed\n",
      "ENSG00000250746.1 completed\n",
      "ENSG00000157426.8 completed\n",
      "ENSG00000178163.3 completed\n",
      "ENSG00000163683.6 completed\n",
      "ENSG00000174799.5 completed\n",
      "ENSG00000109787.8 completed\n",
      "ENSG00000214273.4 completed\n",
      "ENSG00000243185.1 completed\n",
      "ENSG00000164168.3 completed\n",
      "ENSG00000231707.3 completed\n",
      "ENSG00000249741.1 completed\n",
      "ENSG00000038219.8 completed\n",
      "ENSG00000138796.10 completed\n",
      "ENSG00000196656.6 completed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000213608.4 completed\n",
      "ENSG00000248254.1 completed\n",
      "ENSG00000249060.1 completed\n",
      "ENSG00000127418.9 completed\n",
      "ENSG00000154274.10 completed\n",
      "ENSG00000251441.1 completed\n",
      "ENSG00000178177.9 completed\n",
      "ENSG00000163956.6 completed\n",
      "ENSG00000132463.8 completed\n",
      "ENSG00000169288.12 completed\n",
      "ENSG00000087274.11 completed\n",
      "ENSG00000138642.9 completed\n",
      "ENSG00000109686.11 completed\n",
      "ENSG00000237170.3 completed\n",
      "ENSG00000163071.6 completed\n",
      "ENSG00000137628.11 completed\n",
      "ENSG00000168993.9 completed\n",
      "ENSG00000249006.1 completed\n",
      "ENSG00000236562.3 completed\n",
      "ENSG00000164144.10 completed\n",
      "ENSG00000171476.16 completed\n",
      "ENSG00000179299.12 completed\n",
      "ENSG00000244002.1 completed\n",
      "ENSG00000123739.6 completed\n",
      "ENSG00000249072.1 completed\n",
      "ENSG00000170917.8 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000249341.1 completed\n",
      "ENSG00000170801.5 completed\n",
      "ENSG00000162840.3 completed\n",
      "ENSG00000205959.3 completed\n",
      "ENSG00000249272.1 completed\n",
      "ENSG00000249755.1 completed\n",
      "ENSG00000251022.1 completed\n",
      "ENSG00000249685.1 completed\n",
      "ENSG00000118579.6 completed\n",
      "ENSG00000164164.11 completed\n",
      "ENSG00000184178.11 completed\n",
      "ENSG00000151617.10 completed\n",
      "ENSG00000249004.1 completed\n",
      "ENSG00000123933.9 completed\n",
      "ENSG00000198301.7 completed\n",
      "ENSG00000197894.6 completed\n",
      "ENSG00000177432.5 completed\n",
      "ENSG00000145390.6 completed\n",
      "ENSG00000249256.1 completed\n",
      "ENSG00000164125.10 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000242670.1 completed\n",
      "ENSG00000164104.7 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000109220.6 completed\n",
      "ENSG00000109689.10 completed\n",
      "ENSG00000125386.10 completed\n",
      "ENSG00000197386.6 completed\n",
      "ENSG00000142731.6 completed\n",
      "ENSG00000170502.8 completed\n",
      "ENSG00000213332.4 completed\n",
      "ENSG00000180673.9 completed\n",
      "ENSG00000109685.13 completed\n",
      "ENSG00000074211.9 completed\n",
      "ENSG00000236296.2 completed\n",
      "ENSG00000250227.1 completed\n",
      "ENSG00000251017.1 completed\n",
      "ENSG00000163945.10 completed\n",
      "ENSG00000198155.5 completed\n",
      "ENSG00000246375.2 completed\n",
      "ENSG00000185774.9 completed\n",
      "ENSG00000179387.4 completed\n",
      "ENSG00000145246.8 completed\n",
      "ENSG00000251483.1 completed\n",
      "ENSG00000138780.9 completed\n",
      "ENSG00000174780.10 completed\n",
      "ENSG00000250268.2 completed\n",
      "ENSG00000251624.1 completed\n",
      "ENSG00000250254.1 completed\n",
      "ENSG00000250893.1 completed\n",
      "ENSG00000163644.9 completed\n",
      "ENSG00000138756.12 completed\n",
      "ENSG00000178222.8 completed\n",
      "ENSG00000249565.1 completed\n",
      "ENSG00000168824.10 completed\n",
      "ENSG00000164118.8 completed\n",
      "ENSG00000248780.1 completed\n",
      "ENSG00000151458.7 completed\n",
      "ENSG00000072832.9 completed\n",
      "ENSG00000251271.2 completed\n",
      "ENSG00000174123.6 completed\n",
      "ENSG00000169116.7 completed\n",
      "ENSG00000184139.7 completed\n",
      "ENSG00000261668.1 completed\n",
      "ENSG00000224569.3 completed\n",
      "ENSG00000090316.10 completed\n",
      "ENSG00000145220.9 completed\n",
      "ENSG00000164073.5 completed\n",
      "ENSG00000164162.8 completed\n",
      "ENSG00000196951.5 completed\n",
      "ENSG00000178343.4 completed\n",
      "ENSG00000109674.3 completed\n",
      "ENSG00000138771.9 completed\n",
      "ENSG00000174137.8 completed\n",
      "ENSG00000120519.10 completed\n",
      "ENSG00000240167.1 completed\n",
      "ENSG00000170846.11 completed\n",
      "ENSG00000189157.8 completed\n",
      "ENSG00000168772.9 completed\n",
      "ENSG00000121892.10 completed\n",
      "ENSG00000127415.8 completed\n",
      "ENSG00000033178.7 completed\n",
      "ENSG00000251095.1 completed\n",
      "ENSG00000248926.1 completed\n",
      "ENSG00000259959.1 completed\n",
      "ENSG00000038210.8 completed\n",
      "ENSG00000173320.5 completed\n",
      "ENSG00000138792.5 completed\n",
      "ENSG00000118777.6 completed\n",
      "ENSG00000164066.7 completed\n",
      "ENSG00000250902.1 completed\n",
      "ENSG00000152953.8 completed\n",
      "ENSG00000238206.3 completed\n",
      "ENSG00000250381.1 completed\n",
      "ENSG00000249762.1 completed\n",
      "ENSG00000109475.12 completed\n",
      "ENSG00000080493.9 completed\n",
      "ENSG00000164040.10 completed\n",
      "ENSG00000070190.8 completed\n",
      "ENSG00000125089.10 completed\n",
      "ENSG00000251482.1 completed\n",
      "ENSG00000249443.1 completed\n",
      "ENSG00000169762.11 completed\n",
      "ENSG00000109458.4 completed\n",
      "ENSG00000134962.6 completed\n",
      "ENSG00000251056.1 completed\n",
      "ENSG00000236941.3 completed\n",
      "ENSG00000241002.1 completed\n",
      "ENSG00000218422.2 completed\n",
      "ENSG00000256043.2 completed\n",
      "ENSG00000214988.4 completed\n",
      "ENSG00000154608.8 completed\n",
      "ENSG00000231442.3 completed\n",
      "ENSG00000128039.6 completed\n",
      "ENSG00000235606.3 completed\n",
      "ENSG00000260303.1 completed\n",
      "ENSG00000198948.6 completed\n",
      "ENSG00000109586.6 completed\n",
      "ENSG00000184305.9 completed\n",
      "ENSG00000249242.2 completed\n",
      "ENSG00000248256.1 completed\n",
      "ENSG00000153132.7 completed\n",
      "ENSG00000163738.13 completed\n",
      "ENSG00000174749.4 completed\n",
      "ENSG00000255458.1 completed\n",
      "ENSG00000245293.2 completed\n",
      "ENSG00000132467.2 completed\n",
      "ENSG00000251254.1 completed\n",
      "ENSG00000133256.8 completed\n",
      "ENSG00000197701.7 completed\n",
      "ENSG00000109339.14 completed\n",
      "ENSG00000134851.7 completed\n",
      "ENSG00000248583.1 completed\n",
      "ENSG00000065882.9 completed\n",
      "ENSG00000056050.6 completed\n",
      "ENSG00000124406.12 completed\n",
      "ENSG00000150627.9 completed\n",
      "ENSG00000229894.2 completed\n",
      "ENSG00000145362.11 completed\n",
      "ENSG00000248429.1 completed\n",
      "ENSG00000078140.9 completed\n",
      "ENSG00000250725.1 completed\n",
      "ENSG00000249673.1 completed\n",
      "ENSG00000163629.7 completed\n",
      "ENSG00000215323.4 completed\n",
      "ENSG00000177803.7 completed\n",
      "ENSG00000164303.6 completed\n",
      "ENSG00000138769.6 completed\n",
      "ENSG00000242697.2 completed\n",
      "ENSG00000164074.8 completed\n",
      "ENSG00000118762.3 completed\n",
      "ENSG00000164093.10 completed\n",
      "ENSG00000157404.10 completed\n",
      "ENSG00000128050.4 completed\n",
      "ENSG00000247193.2 completed\n",
      "ENSG00000138772.8 completed\n",
      "ENSG00000072840.8 completed\n",
      "ENSG00000109133.8 completed\n",
      "ENSG00000109684.9 completed\n",
      "ENSG00000240005.1 completed\n",
      "ENSG00000168310.5 completed\n",
      "ENSG00000071205.7 completed\n",
      "ENSG00000248283.1 completed\n",
      "ENSG00000173083.9 completed\n",
      "ENSG00000242083.2 completed\n",
      "ENSG00000198589.6 completed\n",
      "ENSG00000197888.2 completed\n",
      "ENSG00000138798.7 completed\n",
      "ENSG00000249717.1 completed\n",
      "ENSG00000248144.1 completed\n",
      "ENSG00000251229.1 completed\n",
      "ENSG00000250076.1 completed\n",
      "ENSG00000109452.7 completed\n",
      "ENSG00000251615.1 completed\n",
      "ENSG00000250892.1 completed\n",
      "ENSG00000236764.3 completed\n",
      "ENSG00000260265.1 completed\n",
      "ENSG00000260404.1 completed\n",
      "ENSG00000138757.10 completed\n",
      "ENSG00000138641.10 completed\n",
      "ENSG00000246560.2 completed\n",
      "ENSG00000183199.6 completed\n",
      "ENSG00000180015.11 completed\n",
      "ENSG00000168421.7 completed\n",
      "ENSG00000138802.6 completed\n",
      "ENSG00000246526.2 completed\n",
      "ENSG00000180610.9 completed\n",
      "ENSG00000260278.1 completed\n",
      "ENSG00000151725.6 completed\n",
      "ENSG00000109265.7 completed\n",
      "ENSG00000225094.2 completed\n",
      "ENSG00000170448.7 completed\n",
      "ENSG00000233110.1 completed\n",
      "ENSG00000109625.13 completed\n",
      "ENSG00000018189.7 completed\n",
      "ENSG00000154553.8 completed\n",
      "ENSG00000138759.12 completed\n",
      "ENSG00000164136.12 completed\n",
      "ENSG00000248625.1 completed\n",
      "ENSG00000251087.2 completed\n",
      "ENSG00000205097.2 completed\n",
      "ENSG00000249617.1 completed\n",
      "ENSG00000173040.8 completed\n",
      "ENSG00000248161.1 completed\n",
      "ENSG00000214367.3 completed\n",
      "ENSG00000249264.1 completed\n",
      "ENSG00000134852.10 completed\n",
      "ENSG00000225405.3 completed\n",
      "ENSG00000163612.10 completed\n",
      "ENSG00000249936.2 completed\n",
      "ENSG00000050730.11 completed\n",
      "ENSG00000138663.4 completed\n",
      "ENSG00000124882.3 completed\n",
      "ENSG00000163291.10 completed\n",
      "ENSG00000174130.7 completed\n",
      "ENSG00000163138.13 completed\n",
      "ENSG00000250326.1 completed\n",
      "ENSG00000251442.1 completed\n",
      "ENSG00000132465.6 completed\n",
      "ENSG00000248795.1 completed\n",
      "ENSG00000249245.1 completed\n",
      "ENSG00000154277.8 completed\n",
      "ENSG00000109320.6 completed\n",
      "ENSG00000197465.8 completed\n",
      "ENSG00000164022.11 completed\n",
      "ENSG00000109158.6 completed\n",
      "ENSG00000249193.1 completed\n",
      "ENSG00000109171.9 completed\n",
      "ENSG00000174473.10 completed\n",
      "ENSG00000249013.1 completed\n",
      "ENSG00000145217.9 completed\n",
      "ENSG00000186777.7 completed\n",
      "ENSG00000248817.1 completed\n",
      "ENSG00000170006.7 completed\n",
      "ENSG00000227304.3 completed\n",
      "ENSG00000073331.12 completed\n",
      "ENSG00000166530.6 completed\n",
      "ENSG00000150625.12 completed\n",
      "ENSG00000163312.6 completed\n",
      "ENSG00000251676.1 completed\n",
      "ENSG00000137462.6 completed\n",
      "ENSG00000250202.1 completed\n",
      "ENSG00000249381.1 completed\n",
      "ENSG00000175920.11 completed\n",
      "ENSG00000183439.5 completed\n",
      "ENSG00000250651.1 completed\n",
      "ENSG00000145247.7 completed\n",
      "ENSG00000241612.1 completed\n",
      "ENSG00000249604.1 completed\n",
      "ENSG00000138795.5 completed\n",
      "ENSG00000138823.7 completed\n",
      "ENSG00000174599.4 completed\n",
      "ENSG00000109680.6 completed\n",
      "ENSG00000035928.10 completed\n",
      "ENSG00000249858.1 completed\n",
      "ENSG00000145375.6 completed\n",
      "ENSG00000248208.1 completed\n",
      "ENSG00000250384.1 completed\n",
      "ENSG00000168491.5 completed\n",
      "ENSG00000163694.10 completed\n",
      "ENSG00000109667.7 completed\n",
      "ENSG00000198856.8 completed\n",
      "ENSG00000109756.4 completed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000121211.3 completed\n",
      "ENSG00000198515.9 completed\n",
      "ENSG00000164109.8 completed\n",
      "ENSG00000240674.1 completed\n",
      "ENSG00000260641.1 completed\n",
      "ENSG00000109189.7 completed\n",
      "ENSG00000163110.10 completed\n",
      "ENSG00000250144.1 completed\n",
      "ENSG00000145216.11 completed\n",
      "ENSG00000109501.8 completed\n",
      "ENSG00000131127.8 completed\n",
      "ENSG00000138755.5 completed\n",
      "ENSG00000198189.6 completed\n",
      "ENSG00000109323.4 completed\n",
      "ENSG00000196810.3 completed\n",
      "ENSG00000227040.3 completed\n",
      "ENSG00000241411.1 completed\n",
      "ENSG00000174343.5 completed\n",
      "ENSG00000234492.3 completed\n",
      "ENSG00000151806.9 completed\n",
      "ENSG00000159674.7 completed\n",
      "ENSG00000243449.2 completed\n",
      "ENSG00000129187.10 completed\n",
      "ENSG00000168214.16 completed\n",
      "ENSG00000249780.1 completed\n",
      "ENSG00000246095.2 completed\n",
      "ENSG00000118804.7 completed\n",
      "ENSG00000138670.10 completed\n",
      "ENSG00000246090.2 completed\n",
      "ENSG00000164024.7 completed\n",
      "ENSG00000260244.1 completed\n",
      "ENSG00000138674.12 completed\n",
      "ENSG00000251600.1 completed\n",
      "ENSG00000129116.13 completed\n",
      "ENSG00000138658.11 completed\n",
      "ENSG00000251464.1 completed\n",
      "ENSG00000237350.1 completed\n",
      "ENSG00000229924.2 completed\n",
      "ENSG00000250163.1 completed\n",
      "ENSG00000232021.2 completed\n",
      "ENSG00000182308.5 completed\n",
      "ENSG00000164037.11 completed\n",
      "ENSG00000241279.1 completed\n",
      "ENSG00000250771.1 completed\n",
      "ENSG00000239735.1 completed\n",
      "ENSG00000145335.11 completed\n",
      "ENSG00000249592.1 completed\n",
      "ENSG00000250403.1 completed\n",
      "ENSG00000249960.1 completed\n",
      "ENSG00000248019.2 completed\n",
      "ENSG00000249465.1 completed\n",
      "ENSG00000130997.11 completed\n",
      "ENSG00000169435.9 completed\n",
      "ENSG00000181004.5 completed\n",
      "ENSG00000249464.1 completed\n",
      "ENSG00000163682.10 completed\n",
      "ENSG00000154447.10 completed\n",
      "ENSG00000248924.1 completed\n",
      "ENSG00000164032.7 completed\n",
      "ENSG00000173542.4 completed\n",
      "ENSG00000152583.7 completed\n",
      "ENSG00000109805.5 completed\n",
      "ENSG00000151552.7 completed\n",
      "ENSG00000151014.4 completed\n",
      "ENSG00000235413.3 completed\n",
      "ENSG00000205129.4 completed\n",
      "ENSG00000087269.10 completed\n",
      "ENSG00000071127.11 completed\n",
      "ENSG00000138646.4 completed\n",
      "ENSG00000164305.12 completed\n",
      "ENSG00000244021.2 completed\n",
      "ENSG00000145414.4 completed\n",
      "ENSG00000128040.6 completed\n",
      "ENSG00000078177.8 completed\n",
      "ENSG00000237868.1 completed\n",
      "ENSG00000109572.8 completed\n",
      "ENSG00000231841.1 completed\n",
      "ENSG00000145388.8 completed\n",
      "ENSG00000109390.7 completed\n",
      "ENSG00000197712.6 completed\n",
      "ENSG00000213480.3 completed\n",
      "ENSG00000237765.2 completed\n",
      "ENSG00000138764.9 completed\n",
      "ENSG00000168884.10 completed\n",
      "ENSG00000047315.10 completed\n",
      "ENSG00000248971.1 completed\n",
      "ENSG00000156136.5 completed\n",
      "ENSG00000163950.7 completed\n",
      "ENSG00000234801.2 completed\n",
      "ENSG00000248305.1 completed\n",
      "ENSG00000251160.1 completed\n",
      "ENSG00000248375.1 completed\n",
      "ENSG00000168826.11 completed\n",
      "ENSG00000118785.8 completed\n",
      "ENSG00000243894.1 completed\n",
      "ENSG00000251111.1 completed\n",
      "ENSG00000168538.10 completed\n",
      "ENSG00000163104.12 completed\n",
      "ENSG00000234036.3 completed\n",
      "ENSG00000138639.13 completed\n",
      "ENSG00000250826.1 completed\n",
      "ENSG00000170365.5 completed\n",
      "ENSG00000138801.4 completed\n",
      "ENSG00000241923.2 completed\n",
      "ENSG00000251411.1 completed\n",
      "ENSG00000245112.2 completed\n",
      "ENSG00000250050.1 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000180613.6 completed\n",
      "ENSG00000250442.1 completed\n",
      "ENSG00000242882.1 completed\n",
      "ENSG00000138744.8 completed\n",
      "ENSG00000242262.1 completed\n",
      "ENSG00000164038.10 completed\n",
      "ENSG00000132466.12 completed\n",
      "ENSG00000138668.13 completed\n",
      "ENSG00000187653.10 completed\n",
      "ENSG00000124399.3 completed\n",
      "ENSG00000155016.12 completed\n",
      "ENSG00000248356.1 completed\n",
      "ENSG00000249635.1 completed\n",
      "ENSG00000163322.9 completed\n",
      "ENSG00000164120.9 completed\n",
      "ENSG00000183423.6 completed\n",
      "ENSG00000164039.10 completed\n",
      "ENSG00000138660.7 completed\n",
      "ENSG00000250966.1 completed\n",
      "ENSG00000145287.6 completed\n",
      "ENSG00000109445.6 completed\n",
      "ENSG00000255072.1 completed\n",
      "ENSG00000002587.5 completed\n",
      "ENSG00000244422.2 completed\n",
      "ENSG00000249311.2 completed\n",
      "ENSG00000248611.1 completed\n",
      "ENSG00000170185.5 completed\n",
      "ENSG00000213493.3 completed\n",
      "ENSG00000179010.10 completed\n",
      "ENSG00000232648.3 completed\n",
      "ENSG00000137460.4 completed\n",
      "ENSG00000125388.14 completed\n",
      "ENSG00000224097.3 completed\n",
      "ENSG00000186234.7 completed\n",
      "ENSG00000171497.4 completed\n",
      "ENSG00000180346.2 completed\n",
      "ENSG00000151726.8 completed\n",
      "ENSG00000245954.2 completed\n",
      "ENSG00000137463.4 completed\n",
      "ENSG00000121897.9 completed\n",
      "ENSG00000163297.12 completed\n",
      "ENSG00000145214.9 completed\n",
      "ENSG00000196526.6 completed\n",
      "ENSG00000138760.4 completed\n",
      "ENSG00000138758.7 completed\n",
      "ENSG00000188993.3 completed\n",
      "ENSG00000249096.1 completed\n",
      "ENSG00000007062.7 completed\n",
      "ENSG00000035720.3 completed\n",
      "ENSG00000248329.1 completed\n",
      "ENSG00000038002.4 completed\n",
      "ENSG00000169026.8 completed\n",
      "ENSG00000254531.1 completed\n",
      "ENSG00000234568.3 completed\n",
      "ENSG00000138767.8 completed\n",
      "ENSG00000163995.13 completed\n",
      "ENSG00000152990.9 completed\n",
      "ENSG00000248880.1 completed\n",
      "ENSG00000109814.7 completed\n",
      "ENSG00000170522.5 completed\n",
      "ENSG00000138768.9 completed\n",
      "ENSG00000169429.6 completed\n",
      "ENSG00000181826.5 completed\n",
      "ENSG00000240459.2 completed\n",
      "ENSG00000248110.1 completed\n",
      "ENSG00000109762.10 completed\n",
      "ENSG00000163631.11 completed\n",
      "ENSG00000227765.3 completed\n",
      "ENSG00000243175.1 completed\n",
      "ENSG00000213331.4 completed\n",
      "ENSG00000245322.2 completed\n",
      "ENSG00000068078.12 completed\n",
      "ENSG00000118816.5 completed\n",
      "ENSG00000249534.1 completed\n",
      "ENSG00000250075.1 completed\n",
      "ENSG00000250486.1 completed\n",
      "ENSG00000168564.5 completed\n",
      "ENSG00000251580.1 completed\n",
      "ENSG00000173966.7 completed\n",
      "ENSG00000137441.7 completed\n",
      "ENSG00000084092.6 completed\n",
      "ENSG00000241103.1 completed\n",
      "ENSG00000163293.7 completed\n",
      "ENSG00000243312.2 completed\n"
     ]
    }
   ],
   "source": [
    "# We will ramp up the process above, performing the analysis on a whole chromosome\n",
    "# I have selected chromosome 14 for our test\n",
    "chromosome = \"14\"\n",
    "chr14_genes = gene_expr[gene_expr['Chr'] == chromosome]['Gene_Symbol']\n",
    "\n",
    "# Initialize a list to store results for each SNP and gene as rows\n",
    "chr14_results_list = []\n",
    "\n",
    "for the_gene in chr4_genes:\n",
    "    gene_info = gene_expr[gene_expr['Gene_Symbol'] == the_gene]\n",
    "    snps_in_window = bim[(bim['chrom'] == chromosome) & \n",
    "                         (bim['pos'] >= int(gene_info['Coord'].iloc[0])-500000) & \n",
    "                         (bim['pos'] <= int(gene_info['Coord'].iloc[0])+500000)]\n",
    "    \n",
    "    # Get the SNP IDs from the filtered SNPs\n",
    "    snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "    # Filter the genotype DataFrame for the SNPs in the window\n",
    "    genotype_data_window = geno_df.loc[:, snp_ids]\n",
    "    \n",
    "    # Filter both dataframes to only include matching IDs\n",
    "    geno_df_filtered = genotype_data_window.loc[matching_ids]\n",
    "    gene_expr_filtered = gene_expr_t.loc[matching_ids]\n",
    "\n",
    "    # Filter the gene expression data for the gene of interest\n",
    "    gene_expr_filtered = gene_expr_t.loc[matching_ids, the_gene]\n",
    "\n",
    "    # Merge the gene expression data with genotype data\n",
    "    merged_df = pd.concat([gene_expr_filtered, geno_df_filtered], axis=1)\n",
    "\n",
    "    # Rename the gene expression column for clarity\n",
    "    merged_df.rename(columns={the_gene: 'gene_expression'}, inplace=True)\n",
    "    \n",
    "    # Ensure gene expression values are numeric\n",
    "    y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "    for snp in geno_df_filtered.columns:\n",
    "        X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "        X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "        # Drop rows with missing or invalid data (NaN)\n",
    "        valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "        # Run the regression only if valid data exists\n",
    "        if len(valid_data) > 0:\n",
    "            model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "\n",
    "            if len(model.pvalues) > 1:\n",
    "                # Append results for this SNP as a row in the results list\n",
    "                chr4_results_list = chr14_results_list.append({\n",
    "                    'Gene': the_gene,\n",
    "                    'SNP': snp,\n",
    "                    'p-value': model.pvalues.iloc[1],\n",
    "                    'beta': model.params.iloc[1],\n",
    "                    'r_squared': model.rsquared\n",
    "                })\n",
    "            else:\n",
    "                print(f\"Insufficient model output for SNP {snp}, skipping.\")\n",
    "                \n",
    "    print(the_gene + ' completed')\n",
    "\n",
    "# Convert the list of results into a DataFrame\n",
    "chr14_results_df = pd.DataFrame(chr14_results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "33fbba9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr14_results_df.to_csv('chr14_eqtl_data.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "cf9e674e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>SNP</th>\n",
       "      <th>p-value</th>\n",
       "      <th>beta</th>\n",
       "      <th>r_squared</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs2799784</td>\n",
       "      <td>0.480146</td>\n",
       "      <td>-0.002291</td>\n",
       "      <td>1.458764e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs1959919</td>\n",
       "      <td>0.812529</td>\n",
       "      <td>-0.000658</td>\n",
       "      <td>1.646931e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs8008487</td>\n",
       "      <td>0.072117</td>\n",
       "      <td>-0.004474</td>\n",
       "      <td>9.425716e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs1959918</td>\n",
       "      <td>0.812529</td>\n",
       "      <td>-0.000658</td>\n",
       "      <td>1.646931e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs1959928</td>\n",
       "      <td>0.012317</td>\n",
       "      <td>-0.009792</td>\n",
       "      <td>1.817704e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165509</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs10148007</td>\n",
       "      <td>0.595599</td>\n",
       "      <td>0.011221</td>\n",
       "      <td>8.245056e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165510</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs1349376</td>\n",
       "      <td>0.548497</td>\n",
       "      <td>0.012643</td>\n",
       "      <td>1.053664e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165511</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs8006417</td>\n",
       "      <td>0.824377</td>\n",
       "      <td>-0.004875</td>\n",
       "      <td>1.442003e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165512</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs8006648</td>\n",
       "      <td>0.995422</td>\n",
       "      <td>-0.000131</td>\n",
       "      <td>9.638331e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165513</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs2167239</td>\n",
       "      <td>0.883855</td>\n",
       "      <td>-0.003469</td>\n",
       "      <td>6.248916e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>165514 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene         SNP   p-value      beta     r_squared\n",
       "0       ENSG00000249212.1   rs2799784  0.480146 -0.002291  1.458764e-03\n",
       "1       ENSG00000249212.1   rs1959919  0.812529 -0.000658  1.646931e-04\n",
       "2       ENSG00000249212.1   rs8008487  0.072117 -0.004474  9.425716e-03\n",
       "3       ENSG00000249212.1   rs1959918  0.812529 -0.000658  1.646931e-04\n",
       "4       ENSG00000249212.1   rs1959928  0.012317 -0.009792  1.817704e-02\n",
       "...                   ...         ...       ...       ...           ...\n",
       "165509  ENSG00000243312.2  rs10148007  0.595599  0.011221  8.245056e-04\n",
       "165510  ENSG00000243312.2   rs1349376  0.548497  0.012643  1.053664e-03\n",
       "165511  ENSG00000243312.2   rs8006417  0.824377 -0.004875  1.442003e-04\n",
       "165512  ENSG00000243312.2   rs8006648  0.995422 -0.000131  9.638331e-08\n",
       "165513  ENSG00000243312.2   rs2167239  0.883855 -0.003469  6.248916e-05\n",
       "\n",
       "[165514 rows x 5 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr14_results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ede824fc",
   "metadata": {},
   "source": [
    "## Creating a PRS from the eQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "b2467dc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Correlation: 0.04776757645646745\n",
      "R-squared: 0.002281741360524614\n"
     ]
    }
   ],
   "source": [
    "# A potential clumping method starts by using plink:\n",
    "'''\n",
    "# Define the command as a list of strings\n",
    "command = [\n",
    "    \"plink\",\n",
    "    \"--bfile\", \"EUR.14.test${gene}\",\n",
    "    \"--clump-p1\", \"0.0001\",\n",
    "    \"--clump-r2\", \"0.1\",\n",
    "    \"--clump-kb\", \"250\",\n",
    "    \"--clump\", \"chr14_eqtl_data.txt\",\n",
    "    \"--clump-snp-field\", \"SNP\",\n",
    "    \"--clump-field\", \"P\",\n",
    "    \"--out\", \"EUR.14.test_clumped\"\n",
    "]\n",
    "\n",
    "# Run the PLINK command\n",
    "subprocess.run(command)\n",
    "'''\n",
    "\n",
    "# Instead I will use a simpler approach:\n",
    "threshold = 0.05\n",
    "selected_snps = chr14_results_df[chr14_results_df['p-value'] <= threshold]\n",
    "selected_snps_list = selected_snps['SNP'].tolist()\n",
    "\n",
    "geno_selected = geno_df[selected_snps_list]\n",
    "\n",
    "gene_id = 'ENSG00000165355.7' \n",
    "gene_expr_selected = gene_expr[gene_expr['Gene_Symbol'] == gene_id].set_index('Gene_Symbol').T\n",
    "\n",
    "# Identify common sample IDs between the two datasets\n",
    "common_samples = gene_expr_selected.index.intersection(geno_df.index)\n",
    "\n",
    "# Subset both gene expression and genotype data to these common samples\n",
    "gene_expr_selected = gene_expr_selected.loc[common_samples]\n",
    "geno_selected = geno_df.loc[common_samples, selected_snps_list]  # Filter for significant SNPs\n",
    "\n",
    "# Calculate PRS\n",
    "prs_values = np.dot(geno_selected.values, selected_snps['beta'].values)\n",
    "\n",
    "# Standardize PRS values and gene expression values\n",
    "prs_scaler = StandardScaler()\n",
    "gene_expr_scaler = StandardScaler()\n",
    "\n",
    "prs_values_scaled = prs_scaler.fit_transform(prs_values.reshape(-1, 1))\n",
    "gene_expr_scaled = gene_expr_scaler.fit_transform(gene_expr_selected.values)\n",
    "\n",
    "# Fit linear regression\n",
    "model = LinearRegression()\n",
    "model.fit(prs_values_scaled, gene_expr_scaled)\n",
    "\n",
    "# Predict and evaluate\n",
    "predicted_expr_scaled = model.predict(prs_values_scaled)\n",
    "\n",
    "# Calculate correlation and R-squared\n",
    "correlation = np.corrcoef(predicted_expr_scaled.flatten(), gene_expr_scaled.flatten())[0,1]\n",
    "r2 = r2_score(gene_expr_scaled, predicted_expr_scaled)\n",
    "\n",
    "print(\"Correlation:\", correlation)\n",
    "print(\"R-squared:\", r2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "fef39f58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using threshold: 1\n",
      "Correlation: 0.04701723549191625\n",
      "R-squared: 0.002210620433302335\n",
      "\n",
      "Using threshold: 0.05\n",
      "Correlation: 0.04776757645646745\n",
      "R-squared: 0.002281741360524614\n",
      "\n",
      "Using threshold: 0.01\n",
      "Correlation: 0.011403420193361961\n",
      "R-squared: 0.00013003799210653533\n",
      "\n",
      "Using threshold: 0.001\n",
      "Correlation: 0.005541453223501891\n",
      "R-squared: 3.070770382829213e-05\n",
      "\n",
      "Using threshold: 1e-05\n",
      "Correlation: 0.05436162121881777\n",
      "R-squared: 0.00295518586153809\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Not that I have tried one p-value I would like to test if other p-values may work better:\n",
    "for threshold in [1, 0.05, 0.01, 0.001, 0.00001]:\n",
    "    selected_snps = chr14_results_df[chr14_results_df['p-value'] <= threshold]\n",
    "    selected_snps_list = selected_snps['SNP'].tolist()\n",
    "\n",
    "    geno_selected = geno_df[selected_snps_list]\n",
    "\n",
    "    gene_id = 'ENSG00000165355.7' \n",
    "    gene_expr_selected = gene_expr[gene_expr['Gene_Symbol'] == gene_id].set_index('Gene_Symbol').T\n",
    "\n",
    "    # Identify common sample IDs between the two datasets\n",
    "    common_samples = gene_expr_selected.index.intersection(geno_df.index)\n",
    "\n",
    "    # Subset both gene expression and genotype data to these common samples\n",
    "    gene_expr_selected = gene_expr_selected.loc[common_samples]\n",
    "    geno_selected = geno_df.loc[common_samples, selected_snps_list]  # Filter for significant SNPs\n",
    "\n",
    "    # Calculate PRS\n",
    "    prs_values = np.dot(geno_selected.values, selected_snps['beta'].values)\n",
    "\n",
    "    # Standardize PRS values and gene expression values\n",
    "    prs_scaler = StandardScaler()\n",
    "    gene_expr_scaler = StandardScaler()\n",
    "\n",
    "    prs_values_scaled = prs_scaler.fit_transform(prs_values.reshape(-1, 1))\n",
    "    gene_expr_scaled = gene_expr_scaler.fit_transform(gene_expr_selected.values)\n",
    "\n",
    "    # Fit linear regression\n",
    "    model = LinearRegression()\n",
    "    model.fit(prs_values_scaled, gene_expr_scaled)\n",
    "\n",
    "    # Predict and evaluate\n",
    "    predicted_expr_scaled = model.predict(prs_values_scaled)\n",
    "\n",
    "    # Calculate correlation and R-squared\n",
    "    correlation = np.corrcoef(predicted_expr_scaled.flatten(), gene_expr_scaled.flatten())[0,1]\n",
    "    r2 = r2_score(gene_expr_scaled, predicted_expr_scaled)\n",
    "\n",
    "    print(\"Using threshold:\", threshold)\n",
    "    print(\"Correlation:\", correlation)\n",
    "    print(\"R-squared:\", r2)\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa84a5e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
