{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3ad0ac3b",
   "metadata": {},
   "source": [
    "## Start with Imports and Loading in the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "32e7e3e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import pandas_plink as pdpl\n",
    "import statsmodels.api as sm\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "import subprocess\n",
    "from sklearn.metrics import r2_score\n",
    "!pip install plink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "6233cf2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files:   0%|          | 0/3 [00:00<?, ?it/s]/home/jhauk/.local/lib/python3.9/site-packages/pandas_plink/_read.py:338: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = read_csv(\n",
      "Mapping files:  33%|███▎      | 1/3 [00:00<00:00,  8.64it/s]/home/jhauk/.local/lib/python3.9/site-packages/pandas_plink/_read.py:338: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = read_csv(\n",
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 17.27it/s]\n"
     ]
    }
   ],
   "source": [
    "# There are three file types for genetic data that interest us\n",
    "# Multiple different samples of these files can be found on 1000 Genomes\n",
    "# In the interest of size and reproducibility we will use just EUR.1 in the beginning\n",
    "bim, fam, G = pdpl.read_plink(\"EUR.1.test/1000G.EUR.1\")\n",
    "\n",
    "# This file contains the gene expression data we will be using\n",
    "gene_expr = pd.read_csv('GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# We also want to grab the gene annotations\n",
    "with gzip.open(\"gene_annot.txt.gz\", \"rt\") as f:\n",
    "    # Read the file content into a pandas DataFrame\n",
    "    gene_annotations = pd.read_csv(f, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c01f34d4",
   "metadata": {},
   "source": [
    "## Data Wrangling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "5229e3a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want to combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df = pd.DataFrame(G.compute().transpose())\n",
    "geno_df.columns = bim['snp']  # Use SNP IDs as columns\n",
    "geno_df.index = fam['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t = gene_expr.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t.index.name = 'iid'\n",
    "matching_ids = gene_expr_t.index.intersection(fam['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = geno_df.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca1e49c3",
   "metadata": {},
   "source": [
    "## Test eQTL Analysis for a Single Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c0b818b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Most Significant Results:\n",
      "            p-value      beta  r_squared\n",
      "rs4970420  0.004872 -0.016493   0.022943\n",
      "rs9442373  0.030045  0.010051   0.013688\n",
      "rs3748594  0.038706 -0.023893   0.012437\n",
      "rs3748593  0.052245 -0.022778   0.010973\n",
      "rs6696281  0.071271 -0.017948   0.009482\n",
      "\n",
      "All Results:\n",
      "             p-value      beta  r_squared\n",
      "rs3094315   0.179067 -0.008622   0.005273\n",
      "rs3131972   0.150053 -0.009219   0.006048\n",
      "rs3131969   0.365582 -0.006137   0.002394\n",
      "rs1048488   0.208763 -0.007981   0.004616\n",
      "rs3115850   0.223496 -0.007717   0.004330\n",
      "...              ...       ...        ...\n",
      "rs1536168   0.104462 -0.018155   0.007689\n",
      "rs3737717   0.942082 -0.001428   0.000015\n",
      "rs12103     0.541356 -0.003543   0.001092\n",
      "rs12142199  0.302081 -0.005895   0.003114\n",
      "rs34054191  0.759153 -0.003741   0.000275\n",
      "\n",
      "[161 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# Randomly select a gene to test:\n",
    "gene_of_interest = 'ENSG00000152931.6'\n",
    "\n",
    "# Adjust window according to gene location\n",
    "# [We will automate this process in the next step]\n",
    "snps_in_window = bim[(bim['chrom'] == \"1\") & \n",
    "                         (bim['pos'] >= 252566) & \n",
    "                         (bim['pos'] <= 1252566)]\n",
    "\n",
    "# Get the SNP IDs from the filtered SNPs\n",
    "snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "# Filter the genotype DataFrame for the SNPs in the window\n",
    "genotype_data_window = geno_df.loc[:, snp_ids]\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = genotype_data_window.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]\n",
    "\n",
    "# Filter the gene expression data for the gene of interest\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids, gene_of_interest]\n",
    "\n",
    "# Merge the gene expression data with genotype data\n",
    "merged_df = pd.concat([gene_expr_filtered, geno_df_filtered], axis=1)\n",
    "\n",
    "# Rename the gene expression column for clarity\n",
    "merged_df.rename(columns={gene_of_interest: 'gene_expression'}, inplace=True)\n",
    "\n",
    "# Ensure gene expression values are numeric\n",
    "y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "# Iterate over each SNP column in the genotype data (independent variables)\n",
    "results = {}\n",
    "for snp in geno_df_filtered.columns:\n",
    "    X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "    X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "    # Drop rows with missing or invalid data (NaN)\n",
    "    valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "    # Run the regression only if valid data exists\n",
    "    if len(valid_data) > 0:\n",
    "        model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "        \n",
    "        # Store the result for this SNP\n",
    "        results[snp] = {\n",
    "            'p-value': model.pvalues.iloc[1],  # Use .iloc to access the p-value for the SNP\n",
    "            'beta': model.params.iloc[1],  # Use .iloc to access the beta coefficient for the SNP\n",
    "            'r_squared': model.rsquared  # R-squared value\n",
    "        }\n",
    "\n",
    "# Convert results to a DataFrame for easier analysis\n",
    "results_df = pd.DataFrame(results).transpose()\n",
    "\n",
    "# Sort by p-value to find significant eQTLs\n",
    "significant_eqtls = results_df.sort_values('p-value').head()\n",
    "print(\"Most Significant Results:\")\n",
    "print(significant_eqtls)\n",
    "print(\"\\nAll Results:\")\n",
    "print(results_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01bc7d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust the results for plotting purposes\n",
    "results_df['neg_log10_p'] = -np.log10(results_df['p-value'])\n",
    "results_df1 = pd.merge(results_df, snps_in_window, left_index=True, right_on='snp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7242dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a locus zoom plot to look at results\n",
    "# Set the aesthetics for the plot\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "# Create the plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(data=results_df1, x='pos', y='neg_log10_p', color='blue', s=100)\n",
    "\n",
    "# Adding labels and title\n",
    "plt.title('LocusZoom Plot of eQTL Analysis', fontsize=16)\n",
    "plt.xlabel('Genomic Coordinate (bp)', fontsize=14)\n",
    "plt.ylabel('-log10(p-value)', fontsize=14)\n",
    "\n",
    "# Adding a horizontal line for significance threshold (e.g., p < 0.05)\n",
    "significance_threshold = -np.log10(0.05)\n",
    "plt.axhline(y=significance_threshold, color='red', linestyle='--', label='p < 0.05')\n",
    "\n",
    "# Customize x-axis limits (optional)\n",
    "plt.xlim([results_df1['pos'].min() - 50000, results_df1['pos'].max() + 50000])\n",
    "\n",
    "# Add legend\n",
    "plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12b3e10f",
   "metadata": {},
   "source": [
    "## Perform cis-eQTL for All the Genes on a Chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "92f6ff0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Mapping files:   0%|          | 0/3 [00:00<?, ?it/s]\u001b[A/home/jhauk/.local/lib/python3.9/site-packages/pandas_plink/_read.py:338: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = read_csv(\n",
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 48.29it/s]\n"
     ]
    }
   ],
   "source": [
    "# We will now repeat some of the process from above to transition to using chromosome 14\n",
    "# This will help us in terms of size and time\n",
    "bim, fam, G = pdpl.read_plink(\"EUR.14.test/1000G.EUR.14\")\n",
    "\n",
    "# This file contains the gene expression data we will be using\n",
    "gene_expr = pd.read_csv('GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz', compression='gzip', sep='\\t')\n",
    "\n",
    "# We also want to grab the gene annotations\n",
    "with gzip.open(\"gene_annot.txt.gz\", \"rt\") as f:\n",
    "    # Read the file content into a pandas DataFrame\n",
    "    gene_annotations = pd.read_csv(f, sep=\"\\t\")\n",
    "    \n",
    "# We want to combine the files so that we have users as rows, and their snps as columns\n",
    "geno_df = pd.DataFrame(G.compute().transpose())\n",
    "geno_df.columns = bim['snp']  # Use SNP IDs as columns\n",
    "geno_df.index = fam['iid']  # Use individual IDs as rows\n",
    "\n",
    "# Transpose gene expression data so that individual IDs are in the index as well\n",
    "gene_expr_t = gene_expr.set_index('Gene_Symbol').transpose()\n",
    "\n",
    "# The data isn't always consistent across all the files, so we want to only pull data we have all the information for\n",
    "# Extract the individual IDs that match those in fam['iid']\n",
    "gene_expr_t.index.name = 'iid'\n",
    "matching_ids = gene_expr_t.index.intersection(fam['iid'])\n",
    "\n",
    "# Filter both dataframes to only include these matching IDs\n",
    "geno_df_filtered = geno_df.loc[matching_ids]\n",
    "gene_expr_filtered = gene_expr_t.loc[matching_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "36a4ae92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000243417.1 completed\n",
      "ENSG00000152969.11 completed\n",
      "ENSG00000249212.1 completed\n",
      "ENSG00000251297.1 completed\n",
      "ENSG00000248155.1 completed\n",
      "ENSG00000178950.11 completed\n",
      "ENSG00000145365.8 completed\n",
      "ENSG00000164096.10 completed\n",
      "ENSG00000186222.3 completed\n",
      "ENSG00000248049.2 completed\n",
      "ENSG00000164167.5 completed\n",
      "ENSG00000168785.3 completed\n",
      "ENSG00000225892.3 completed\n",
      "ENSG00000249952.1 completed\n",
      "ENSG00000197410.6 completed\n",
      "ENSG00000251555.1 completed\n",
      "ENSG00000248697.1 completed\n",
      "ENSG00000152785.5 completed\n",
      "ENSG00000151470.7 completed\n",
      "ENSG00000241281.1 completed\n",
      "ENSG00000185049.7 completed\n",
      "ENSG00000250910.1 completed\n",
      "ENSG00000223305.1 completed\n",
      "ENSG00000109163.6 completed\n",
      "ENSG00000213522.3 completed\n",
      "ENSG00000151718.11 completed\n",
      "ENSG00000061918.8 completed\n",
      "ENSG00000186212.2 completed\n",
      "ENSG00000182903.11 completed\n",
      "ENSG00000251587.1 completed\n",
      "ENSG00000157796.12 completed\n",
      "ENSG00000236530.1 completed\n",
      "ENSG00000087008.10 completed\n",
      "ENSG00000248394.1 completed\n",
      "ENSG00000213492.2 completed\n",
      "ENSG00000163281.6 completed\n",
      "ENSG00000248632.1 completed\n",
      "ENSG00000164056.6 completed\n",
      "ENSG00000250666.1 completed\n",
      "ENSG00000109270.8 completed\n",
      "ENSG00000169245.4 completed\n",
      "ENSG00000250312.2 completed\n",
      "ENSG00000249348.1 completed\n",
      "ENSG00000244004.1 completed\n",
      "ENSG00000151729.5 completed\n",
      "ENSG00000249502.1 completed\n",
      "ENSG00000249241.1 completed\n",
      "ENSG00000250030.1 completed\n",
      "ENSG00000164105.3 completed\n",
      "ENSG00000174227.10 completed\n",
      "ENSG00000132406.7 completed\n",
      "ENSG00000173011.10 completed\n",
      "ENSG00000172955.13 completed\n",
      "ENSG00000109794.9 completed\n",
      "ENSG00000138777.15 completed\n",
      "ENSG00000251596.1 completed\n",
      "ENSG00000239983.1 completed\n",
      "ENSG00000168769.7 completed\n",
      "ENSG00000224207.2 completed\n",
      "ENSG00000083857.8 completed\n",
      "ENSG00000249863.1 completed\n",
      "ENSG00000109670.8 completed\n",
      "ENSG00000251332.1 completed\n",
      "ENSG00000250656.1 completed\n",
      "ENSG00000251571.1 completed\n",
      "ENSG00000205595.3 completed\n",
      "ENSG00000137601.11 completed\n",
      "ENSG00000109180.10 completed\n",
      "ENSG00000260296.1 completed\n",
      "ENSG00000236699.3 completed\n",
      "ENSG00000168924.9 completed\n",
      "ENSG00000250626.2 completed\n",
      "ENSG00000075539.9 completed\n",
      "ENSG00000185477.3 completed\n",
      "ENSG00000248645.1 completed\n",
      "ENSG00000197057.4 completed\n",
      "ENSG00000181877.3 completed\n",
      "ENSG00000188848.10 completed\n",
      "ENSG00000164161.5 completed\n",
      "ENSG00000186352.4 completed\n",
      "ENSG00000250220.1 completed\n",
      "ENSG00000224807.5 completed\n",
      "ENSG00000109819.4 completed\n",
      "ENSG00000138778.6 completed\n",
      "ENSG00000172007.5 completed\n",
      "ENSG00000242640.1 completed\n",
      "ENSG00000121895.7 completed\n",
      "ENSG00000263327.1 completed\n",
      "ENSG00000250471.1 completed\n",
      "ENSG00000164070.6 completed\n",
      "ENSG00000164134.6 completed\n",
      "ENSG00000164111.9 completed\n",
      "ENSG00000164323.8 completed\n",
      "ENSG00000251230.1 completed\n",
      "ENSG00000198498.5 completed\n",
      "ENSG00000174720.11 completed\n",
      "ENSG00000251463.1 completed\n",
      "ENSG00000245562.2 completed\n",
      "ENSG00000013288.4 completed\n",
      "ENSG00000205794.3 completed\n",
      "ENSG00000145337.4 completed\n",
      "ENSG00000109534.12 completed\n",
      "ENSG00000109618.7 completed\n",
      "ENSG00000260526.1 completed\n",
      "ENSG00000174796.8 completed\n",
      "ENSG00000145244.7 completed\n",
      "ENSG00000247950.2 completed\n",
      "ENSG00000170871.7 completed\n",
      "ENSG00000164342.6 completed\n",
      "ENSG00000163069.8 completed\n",
      "ENSG00000138821.8 completed\n",
      "ENSG00000260091.1 completed\n",
      "ENSG00000178988.10 completed\n",
      "ENSG00000137449.10 completed\n",
      "ENSG00000248113.1 completed\n",
      "ENSG00000053900.6 completed\n",
      "ENSG00000247708.2 completed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files:   0%|          | 0/3 [01:46<?, ?it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000227725.2 completed\n",
      "ENSG00000138722.5 completed\n",
      "ENSG00000074966.6 completed\n",
      "ENSG00000249934.1 completed\n",
      "ENSG00000169020.5 completed\n",
      "ENSG00000181359.5 completed\n",
      "ENSG00000214980.4 completed\n",
      "ENSG00000163257.6 completed\n",
      "ENSG00000156234.7 completed\n",
      "ENSG00000251432.2 completed\n",
      "ENSG00000249844.1 completed\n",
      "ENSG00000109775.6 completed\n",
      "ENSG00000251259.1 completed\n",
      "ENSG00000249244.1 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000072201.8 completed\n",
      "ENSG00000178458.4 completed\n",
      "ENSG00000163743.9 completed\n",
      "ENSG00000152795.12 completed\n",
      "ENSG00000174607.6 completed\n",
      "ENSG00000242358.1 completed\n",
      "ENSG00000213434.2 completed\n",
      "ENSG00000250131.1 completed\n",
      "ENSG00000132405.14 completed\n",
      "ENSG00000109436.7 completed\n",
      "ENSG00000183432.6 completed\n",
      "ENSG00000232327.3 completed\n",
      "ENSG00000174125.3 completed\n",
      "ENSG00000250413.1 completed\n",
      "ENSG00000169989.2 completed\n",
      "ENSG00000145349.12 completed\n",
      "ENSG00000250980.1 completed\n",
      "ENSG00000258013.1 completed\n",
      "ENSG00000164031.12 completed\n",
      "ENSG00000085871.4 completed\n",
      "ENSG00000013810.13 completed\n",
      "ENSG00000250327.1 completed\n",
      "ENSG00000181982.11 completed\n",
      "ENSG00000163132.6 completed\n",
      "ENSG00000171234.9 completed\n",
      "ENSG00000180712.3 completed\n",
      "ENSG00000153147.4 completed\n",
      "ENSG00000249896.1 completed\n",
      "ENSG00000250566.1 completed\n",
      "ENSG00000138709.13 completed\n",
      "ENSG00000235043.2 completed\n",
      "ENSG00000121210.11 completed\n",
      "ENSG00000189184.7 completed\n",
      "ENSG00000163625.11 completed\n",
      "ENSG00000152784.9 completed\n",
      "ENSG00000052802.8 completed\n",
      "ENSG00000163319.6 completed\n",
      "ENSG00000249239.1 completed\n",
      "ENSG00000145241.6 completed\n",
      "ENSG00000251308.1 completed\n",
      "ENSG00000241877.1 completed\n",
      "ENSG00000225455.3 completed\n",
      "ENSG00000244146.1 completed\n",
      "ENSG00000005059.11 completed\n",
      "ENSG00000249207.1 completed\n",
      "ENSG00000109381.13 completed\n",
      "ENSG00000004468.6 completed\n",
      "ENSG00000047662.4 completed\n",
      "ENSG00000188981.6 completed\n",
      "ENSG00000250033.1 completed\n",
      "ENSG00000145476.10 completed\n",
      "ENSG00000109184.10 completed\n",
      "ENSG00000168556.5 completed\n",
      "ENSG00000248213.1 completed\n",
      "ENSG00000109790.11 completed\n",
      "ENSG00000214846.4 completed\n",
      "ENSG00000169299.8 completed\n",
      "ENSG00000164124.6 completed\n",
      "ENSG00000135220.6 completed\n",
      "ENSG00000168228.9 completed\n",
      "ENSG00000155275.14 completed\n",
      "ENSG00000109332.15 completed\n",
      "ENSG00000084093.10 completed\n",
      "ENSG00000170088.7 completed\n",
      "ENSG00000151612.10 completed\n",
      "ENSG00000261490.1 completed\n",
      "ENSG00000128059.4 completed\n",
      "ENSG00000241651.3 completed\n",
      "ENSG00000249920.1 completed\n",
      "ENSG00000138814.12 completed\n",
      "ENSG00000251284.1 completed\n",
      "ENSG00000213851.3 completed\n",
      "ENSG00000248401.1 completed\n",
      "ENSG00000214062.5 completed\n",
      "ENSG00000171566.7 completed\n",
      "ENSG00000249679.1 completed\n",
      "ENSG00000151466.7 completed\n",
      "ENSG00000226950.2 completed\n",
      "ENSG00000159733.9 completed\n",
      "ENSG00000083896.7 completed\n",
      "ENSG00000138686.5 completed\n",
      "ENSG00000091490.6 completed\n",
      "ENSG00000163697.11 completed\n",
      "ENSG00000249356.1 completed\n",
      "ENSG00000244538.1 completed\n",
      "ENSG00000245748.1 completed\n",
      "ENSG00000059691.7 completed\n",
      "ENSG00000245213.2 completed\n",
      "ENSG00000172403.6 completed\n",
      "ENSG00000164306.6 completed\n",
      "ENSG00000250325.1 completed\n",
      "ENSG00000164163.6 completed\n",
      "ENSG00000002549.7 completed\n",
      "ENSG00000173013.4 completed\n",
      "ENSG00000242169.1 completed\n",
      "ENSG00000240669.1 completed\n",
      "ENSG00000168843.9 completed\n",
      "ENSG00000174792.6 completed\n",
      "ENSG00000218336.3 completed\n",
      "ENSG00000227118.3 completed\n",
      "ENSG00000249509.1 completed\n",
      "ENSG00000052795.8 completed\n",
      "ENSG00000205940.5 completed\n",
      "ENSG00000014824.7 completed\n",
      "ENSG00000153064.7 completed\n",
      "ENSG00000138640.10 completed\n",
      "ENSG00000231160.3 completed\n",
      "ENSG00000145416.9 completed\n",
      "ENSG00000168743.8 completed\n",
      "ENSG00000109466.9 completed\n",
      "ENSG00000243199.1 completed\n",
      "ENSG00000245958.2 completed\n",
      "ENSG00000138785.10 completed\n",
      "ENSG00000251313.1 completed\n",
      "ENSG00000151247.8 completed\n",
      "ENSG00000145293.9 completed\n",
      "ENSG00000109743.6 completed\n",
      "ENSG00000251669.1 completed\n",
      "ENSG00000163633.6 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000134853.7 completed\n",
      "ENSG00000109536.6 completed\n",
      "ENSG00000251429.1 completed\n",
      "ENSG00000138794.5 completed\n",
      "ENSG00000150961.10 completed\n",
      "ENSG00000138688.9 completed\n",
      "ENSG00000145439.7 completed\n",
      "ENSG00000248122.1 completed\n",
      "ENSG00000226439.3 completed\n",
      "ENSG00000250896.1 completed\n",
      "ENSG00000250317.3 completed\n",
      "ENSG00000127419.10 completed\n",
      "ENSG00000154556.13 completed\n",
      "ENSG00000168818.5 completed\n",
      "ENSG00000249138.1 completed\n",
      "ENSG00000248336.1 completed\n",
      "ENSG00000179979.6 completed\n",
      "ENSG00000123737.8 completed\n",
      "ENSG00000164142.11 completed\n",
      "ENSG00000248571.1 completed\n",
      "ENSG00000185818.7 completed\n",
      "ENSG00000249548.1 completed\n",
      "ENSG00000248988.1 completed\n",
      "ENSG00000173085.9 completed\n",
      "ENSG00000251595.3 completed\n",
      "ENSG00000145386.5 completed\n",
      "ENSG00000145248.6 completed\n",
      "ENSG00000234841.2 completed\n",
      "ENSG00000213406.3 completed\n",
      "ENSG00000164114.13 completed\n",
      "ENSG00000129128.8 completed\n",
      "ENSG00000121089.4 completed\n",
      "ENSG00000249056.1 completed\n",
      "ENSG00000215375.2 completed\n",
      "ENSG00000163626.12 completed\n",
      "ENSG00000250768.1 completed\n",
      "ENSG00000145391.7 completed\n",
      "ENSG00000172493.15 completed\n",
      "ENSG00000234386.3 completed\n",
      "ENSG00000248866.1 completed\n",
      "ENSG00000181381.9 completed\n",
      "ENSG00000250412.1 completed\n",
      "ENSG00000178403.3 completed\n",
      "ENSG00000248774.1 completed\n",
      "ENSG00000245685.2 completed\n",
      "ENSG00000196620.4 completed\n",
      "ENSG00000047365.6 completed\n",
      "ENSG00000169019.9 completed\n",
      "ENSG00000254360.1 completed\n",
      "ENSG00000229204.2 completed\n",
      "ENSG00000243280.2 completed\n",
      "ENSG00000138735.11 completed\n",
      "ENSG00000109771.11 completed\n",
      "ENSG00000249532.1 completed\n",
      "ENSG00000184985.12 completed\n",
      "ENSG00000090989.13 completed\n",
      "ENSG00000151790.4 completed\n",
      "ENSG00000164116.12 completed\n",
      "ENSG00000159788.13 completed\n",
      "ENSG00000248546.3 completed\n",
      "ENSG00000151012.9 completed\n",
      "ENSG00000213370.3 completed\n",
      "ENSG00000145358.2 completed\n",
      "ENSG00000250610.1 completed\n",
      "ENSG00000171503.7 completed\n",
      "ENSG00000087266.11 completed\n",
      "ENSG00000250950.1 completed\n",
      "ENSG00000138750.10 completed\n",
      "ENSG00000109519.7 completed\n",
      "ENSG00000229450.2 completed\n",
      "ENSG00000234125.3 completed\n",
      "ENSG00000164169.8 completed\n",
      "ENSG00000145284.7 completed\n",
      "ENSG00000228981.3 completed\n",
      "ENSG00000248890.1 completed\n",
      "ENSG00000249700.1 completed\n",
      "ENSG00000109654.10 completed\n",
      "ENSG00000063978.10 completed\n",
      "ENSG00000145348.11 completed\n",
      "ENSG00000128045.5 completed\n",
      "ENSG00000157869.10 completed\n",
      "ENSG00000153130.12 completed\n",
      "ENSG00000168152.7 completed\n",
      "ENSG00000077684.10 completed\n",
      "ENSG00000189308.6 completed\n",
      "ENSG00000248196.1 completed\n",
      "ENSG00000145354.5 completed\n",
      "ENSG00000164099.3 completed\n",
      "ENSG00000196782.7 completed\n",
      "ENSG00000109736.10 completed\n",
      "ENSG00000251288.1 completed\n",
      "ENSG00000109606.7 completed\n",
      "ENSG00000250215.1 completed\n",
      "ENSG00000248785.1 completed\n",
      "ENSG00000248180.1 completed\n",
      "ENSG00000145331.9 completed\n",
      "ENSG00000138685.7 completed\n",
      "ENSG00000164023.10 completed\n",
      "ENSG00000261643.1 completed\n",
      "ENSG00000244459.2 completed\n",
      "ENSG00000159692.11 completed\n",
      "ENSG00000248778.1 completed\n",
      "ENSG00000251455.1 completed\n",
      "ENSG00000109321.6 completed\n",
      "ENSG00000241741.1 completed\n",
      "ENSG00000230069.3 completed\n",
      "ENSG00000138678.6 completed\n",
      "ENSG00000151611.8 completed\n",
      "ENSG00000170509.7 completed\n",
      "ENSG00000138698.9 completed\n",
      "ENSG00000251333.2 completed\n",
      "ENSG00000132464.7 completed\n",
      "ENSG00000205100.2 completed\n",
      "ENSG00000248641.1 completed\n",
      "ENSG00000179967.10 completed\n",
      "ENSG00000250185.1 completed\n",
      "ENSG00000151623.9 completed\n",
      "ENSG00000250568.1 completed\n",
      "ENSG00000185619.13 completed\n",
      "ENSG00000213448.3 completed\n",
      "ENSG00000178972.3 completed\n",
      "ENSG00000248590.2 completed\n",
      "ENSG00000249823.1 completed\n",
      "ENSG00000253917.2 completed\n",
      "ENSG00000168936.6 completed\n",
      "ENSG00000250410.1 completed\n",
      "ENSG00000048342.10 completed\n",
      "ENSG00000145425.5 completed\n",
      "ENSG00000182552.10 completed\n",
      "ENSG00000249334.1 completed\n",
      "ENSG00000260519.1 completed\n",
      "ENSG00000118564.10 completed\n",
      "ENSG00000248115.1 completed\n",
      "ENSG00000248327.1 completed\n",
      "ENSG00000135605.7 completed\n",
      "ENSG00000205208.4 completed\n",
      "ENSG00000256143.1 completed\n",
      "ENSG00000250321.1 completed\n",
      "ENSG00000145332.8 completed\n",
      "ENSG00000164117.7 completed\n",
      "ENSG00000249347.1 completed\n",
      "ENSG00000250746.1 completed\n",
      "ENSG00000157426.8 completed\n",
      "ENSG00000178163.3 completed\n",
      "ENSG00000163683.6 completed\n",
      "ENSG00000174799.5 completed\n",
      "ENSG00000109787.8 completed\n",
      "ENSG00000214273.4 completed\n",
      "ENSG00000243185.1 completed\n",
      "ENSG00000164168.3 completed\n",
      "ENSG00000231707.3 completed\n",
      "ENSG00000249741.1 completed\n",
      "ENSG00000038219.8 completed\n",
      "ENSG00000138796.10 completed\n",
      "ENSG00000196656.6 completed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000213608.4 completed\n",
      "ENSG00000248254.1 completed\n",
      "ENSG00000249060.1 completed\n",
      "ENSG00000127418.9 completed\n",
      "ENSG00000154274.10 completed\n",
      "ENSG00000251441.1 completed\n",
      "ENSG00000178177.9 completed\n",
      "ENSG00000163956.6 completed\n",
      "ENSG00000132463.8 completed\n",
      "ENSG00000169288.12 completed\n",
      "ENSG00000087274.11 completed\n",
      "ENSG00000138642.9 completed\n",
      "ENSG00000109686.11 completed\n",
      "ENSG00000237170.3 completed\n",
      "ENSG00000163071.6 completed\n",
      "ENSG00000137628.11 completed\n",
      "ENSG00000168993.9 completed\n",
      "ENSG00000249006.1 completed\n",
      "ENSG00000236562.3 completed\n",
      "ENSG00000164144.10 completed\n",
      "ENSG00000171476.16 completed\n",
      "ENSG00000179299.12 completed\n",
      "ENSG00000244002.1 completed\n",
      "ENSG00000123739.6 completed\n",
      "ENSG00000249072.1 completed\n",
      "ENSG00000170917.8 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000249341.1 completed\n",
      "ENSG00000170801.5 completed\n",
      "ENSG00000162840.3 completed\n",
      "ENSG00000205959.3 completed\n",
      "ENSG00000249272.1 completed\n",
      "ENSG00000249755.1 completed\n",
      "ENSG00000251022.1 completed\n",
      "ENSG00000249685.1 completed\n",
      "ENSG00000118579.6 completed\n",
      "ENSG00000164164.11 completed\n",
      "ENSG00000184178.11 completed\n",
      "ENSG00000151617.10 completed\n",
      "ENSG00000249004.1 completed\n",
      "ENSG00000123933.9 completed\n",
      "ENSG00000198301.7 completed\n",
      "ENSG00000197894.6 completed\n",
      "ENSG00000177432.5 completed\n",
      "ENSG00000145390.6 completed\n",
      "ENSG00000249256.1 completed\n",
      "ENSG00000164125.10 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000242670.1 completed\n",
      "ENSG00000164104.7 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000109220.6 completed\n",
      "ENSG00000109689.10 completed\n",
      "ENSG00000125386.10 completed\n",
      "ENSG00000197386.6 completed\n",
      "ENSG00000142731.6 completed\n",
      "ENSG00000170502.8 completed\n",
      "ENSG00000213332.4 completed\n",
      "ENSG00000180673.9 completed\n",
      "ENSG00000109685.13 completed\n",
      "ENSG00000074211.9 completed\n",
      "ENSG00000236296.2 completed\n",
      "ENSG00000250227.1 completed\n",
      "ENSG00000251017.1 completed\n",
      "ENSG00000163945.10 completed\n",
      "ENSG00000198155.5 completed\n",
      "ENSG00000246375.2 completed\n",
      "ENSG00000185774.9 completed\n",
      "ENSG00000179387.4 completed\n",
      "ENSG00000145246.8 completed\n",
      "ENSG00000251483.1 completed\n",
      "ENSG00000138780.9 completed\n",
      "ENSG00000174780.10 completed\n",
      "ENSG00000250268.2 completed\n",
      "ENSG00000251624.1 completed\n",
      "ENSG00000250254.1 completed\n",
      "ENSG00000250893.1 completed\n",
      "ENSG00000163644.9 completed\n",
      "ENSG00000138756.12 completed\n",
      "ENSG00000178222.8 completed\n",
      "ENSG00000249565.1 completed\n",
      "ENSG00000168824.10 completed\n",
      "ENSG00000164118.8 completed\n",
      "ENSG00000248780.1 completed\n",
      "ENSG00000151458.7 completed\n",
      "ENSG00000072832.9 completed\n",
      "ENSG00000251271.2 completed\n",
      "ENSG00000174123.6 completed\n",
      "ENSG00000169116.7 completed\n",
      "ENSG00000184139.7 completed\n",
      "ENSG00000261668.1 completed\n",
      "ENSG00000224569.3 completed\n",
      "ENSG00000090316.10 completed\n",
      "ENSG00000145220.9 completed\n",
      "ENSG00000164073.5 completed\n",
      "ENSG00000164162.8 completed\n",
      "ENSG00000196951.5 completed\n",
      "ENSG00000178343.4 completed\n",
      "ENSG00000109674.3 completed\n",
      "ENSG00000138771.9 completed\n",
      "ENSG00000174137.8 completed\n",
      "ENSG00000120519.10 completed\n",
      "ENSG00000240167.1 completed\n",
      "ENSG00000170846.11 completed\n",
      "ENSG00000189157.8 completed\n",
      "ENSG00000168772.9 completed\n",
      "ENSG00000121892.10 completed\n",
      "ENSG00000127415.8 completed\n",
      "ENSG00000033178.7 completed\n",
      "ENSG00000251095.1 completed\n",
      "ENSG00000248926.1 completed\n",
      "ENSG00000259959.1 completed\n",
      "ENSG00000038210.8 completed\n",
      "ENSG00000173320.5 completed\n",
      "ENSG00000138792.5 completed\n",
      "ENSG00000118777.6 completed\n",
      "ENSG00000164066.7 completed\n",
      "ENSG00000250902.1 completed\n",
      "ENSG00000152953.8 completed\n",
      "ENSG00000238206.3 completed\n",
      "ENSG00000250381.1 completed\n",
      "ENSG00000249762.1 completed\n",
      "ENSG00000109475.12 completed\n",
      "ENSG00000080493.9 completed\n",
      "ENSG00000164040.10 completed\n",
      "ENSG00000070190.8 completed\n",
      "ENSG00000125089.10 completed\n",
      "ENSG00000251482.1 completed\n",
      "ENSG00000249443.1 completed\n",
      "ENSG00000169762.11 completed\n",
      "ENSG00000109458.4 completed\n",
      "ENSG00000134962.6 completed\n",
      "ENSG00000251056.1 completed\n",
      "ENSG00000236941.3 completed\n",
      "ENSG00000241002.1 completed\n",
      "ENSG00000218422.2 completed\n",
      "ENSG00000256043.2 completed\n",
      "ENSG00000214988.4 completed\n",
      "ENSG00000154608.8 completed\n",
      "ENSG00000231442.3 completed\n",
      "ENSG00000128039.6 completed\n",
      "ENSG00000235606.3 completed\n",
      "ENSG00000260303.1 completed\n",
      "ENSG00000198948.6 completed\n",
      "ENSG00000109586.6 completed\n",
      "ENSG00000184305.9 completed\n",
      "ENSG00000249242.2 completed\n",
      "ENSG00000248256.1 completed\n",
      "ENSG00000153132.7 completed\n",
      "ENSG00000163738.13 completed\n",
      "ENSG00000174749.4 completed\n",
      "ENSG00000255458.1 completed\n",
      "ENSG00000245293.2 completed\n",
      "ENSG00000132467.2 completed\n",
      "ENSG00000251254.1 completed\n",
      "ENSG00000133256.8 completed\n",
      "ENSG00000197701.7 completed\n",
      "ENSG00000109339.14 completed\n",
      "ENSG00000134851.7 completed\n",
      "ENSG00000248583.1 completed\n",
      "ENSG00000065882.9 completed\n",
      "ENSG00000056050.6 completed\n",
      "ENSG00000124406.12 completed\n",
      "ENSG00000150627.9 completed\n",
      "ENSG00000229894.2 completed\n",
      "ENSG00000145362.11 completed\n",
      "ENSG00000248429.1 completed\n",
      "ENSG00000078140.9 completed\n",
      "ENSG00000250725.1 completed\n",
      "ENSG00000249673.1 completed\n",
      "ENSG00000163629.7 completed\n",
      "ENSG00000215323.4 completed\n",
      "ENSG00000177803.7 completed\n",
      "ENSG00000164303.6 completed\n",
      "ENSG00000138769.6 completed\n",
      "ENSG00000242697.2 completed\n",
      "ENSG00000164074.8 completed\n",
      "ENSG00000118762.3 completed\n",
      "ENSG00000164093.10 completed\n",
      "ENSG00000157404.10 completed\n",
      "ENSG00000128050.4 completed\n",
      "ENSG00000247193.2 completed\n",
      "ENSG00000138772.8 completed\n",
      "ENSG00000072840.8 completed\n",
      "ENSG00000109133.8 completed\n",
      "ENSG00000109684.9 completed\n",
      "ENSG00000240005.1 completed\n",
      "ENSG00000168310.5 completed\n",
      "ENSG00000071205.7 completed\n",
      "ENSG00000248283.1 completed\n",
      "ENSG00000173083.9 completed\n",
      "ENSG00000242083.2 completed\n",
      "ENSG00000198589.6 completed\n",
      "ENSG00000197888.2 completed\n",
      "ENSG00000138798.7 completed\n",
      "ENSG00000249717.1 completed\n",
      "ENSG00000248144.1 completed\n",
      "ENSG00000251229.1 completed\n",
      "ENSG00000250076.1 completed\n",
      "ENSG00000109452.7 completed\n",
      "ENSG00000251615.1 completed\n",
      "ENSG00000250892.1 completed\n",
      "ENSG00000236764.3 completed\n",
      "ENSG00000260265.1 completed\n",
      "ENSG00000260404.1 completed\n",
      "ENSG00000138757.10 completed\n",
      "ENSG00000138641.10 completed\n",
      "ENSG00000246560.2 completed\n",
      "ENSG00000183199.6 completed\n",
      "ENSG00000180015.11 completed\n",
      "ENSG00000168421.7 completed\n",
      "ENSG00000138802.6 completed\n",
      "ENSG00000246526.2 completed\n",
      "ENSG00000180610.9 completed\n",
      "ENSG00000260278.1 completed\n",
      "ENSG00000151725.6 completed\n",
      "ENSG00000109265.7 completed\n",
      "ENSG00000225094.2 completed\n",
      "ENSG00000170448.7 completed\n",
      "ENSG00000233110.1 completed\n",
      "ENSG00000109625.13 completed\n",
      "ENSG00000018189.7 completed\n",
      "ENSG00000154553.8 completed\n",
      "ENSG00000138759.12 completed\n",
      "ENSG00000164136.12 completed\n",
      "ENSG00000248625.1 completed\n",
      "ENSG00000251087.2 completed\n",
      "ENSG00000205097.2 completed\n",
      "ENSG00000249617.1 completed\n",
      "ENSG00000173040.8 completed\n",
      "ENSG00000248161.1 completed\n",
      "ENSG00000214367.3 completed\n",
      "ENSG00000249264.1 completed\n",
      "ENSG00000134852.10 completed\n",
      "ENSG00000225405.3 completed\n",
      "ENSG00000163612.10 completed\n",
      "ENSG00000249936.2 completed\n",
      "ENSG00000050730.11 completed\n",
      "ENSG00000138663.4 completed\n",
      "ENSG00000124882.3 completed\n",
      "ENSG00000163291.10 completed\n",
      "ENSG00000174130.7 completed\n",
      "ENSG00000163138.13 completed\n",
      "ENSG00000250326.1 completed\n",
      "ENSG00000251442.1 completed\n",
      "ENSG00000132465.6 completed\n",
      "ENSG00000248795.1 completed\n",
      "ENSG00000249245.1 completed\n",
      "ENSG00000154277.8 completed\n",
      "ENSG00000109320.6 completed\n",
      "ENSG00000197465.8 completed\n",
      "ENSG00000164022.11 completed\n",
      "ENSG00000109158.6 completed\n",
      "ENSG00000249193.1 completed\n",
      "ENSG00000109171.9 completed\n",
      "ENSG00000174473.10 completed\n",
      "ENSG00000249013.1 completed\n",
      "ENSG00000145217.9 completed\n",
      "ENSG00000186777.7 completed\n",
      "ENSG00000248817.1 completed\n",
      "ENSG00000170006.7 completed\n",
      "ENSG00000227304.3 completed\n",
      "ENSG00000073331.12 completed\n",
      "ENSG00000166530.6 completed\n",
      "ENSG00000150625.12 completed\n",
      "ENSG00000163312.6 completed\n",
      "ENSG00000251676.1 completed\n",
      "ENSG00000137462.6 completed\n",
      "ENSG00000250202.1 completed\n",
      "ENSG00000249381.1 completed\n",
      "ENSG00000175920.11 completed\n",
      "ENSG00000183439.5 completed\n",
      "ENSG00000250651.1 completed\n",
      "ENSG00000145247.7 completed\n",
      "ENSG00000241612.1 completed\n",
      "ENSG00000249604.1 completed\n",
      "ENSG00000138795.5 completed\n",
      "ENSG00000138823.7 completed\n",
      "ENSG00000174599.4 completed\n",
      "ENSG00000109680.6 completed\n",
      "ENSG00000035928.10 completed\n",
      "ENSG00000249858.1 completed\n",
      "ENSG00000145375.6 completed\n",
      "ENSG00000248208.1 completed\n",
      "ENSG00000250384.1 completed\n",
      "ENSG00000168491.5 completed\n",
      "ENSG00000163694.10 completed\n",
      "ENSG00000109667.7 completed\n",
      "ENSG00000198856.8 completed\n",
      "ENSG00000109756.4 completed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000121211.3 completed\n",
      "ENSG00000198515.9 completed\n",
      "ENSG00000164109.8 completed\n",
      "ENSG00000240674.1 completed\n",
      "ENSG00000260641.1 completed\n",
      "ENSG00000109189.7 completed\n",
      "ENSG00000163110.10 completed\n",
      "ENSG00000250144.1 completed\n",
      "ENSG00000145216.11 completed\n",
      "ENSG00000109501.8 completed\n",
      "ENSG00000131127.8 completed\n",
      "ENSG00000138755.5 completed\n",
      "ENSG00000198189.6 completed\n",
      "ENSG00000109323.4 completed\n",
      "ENSG00000196810.3 completed\n",
      "ENSG00000227040.3 completed\n",
      "ENSG00000241411.1 completed\n",
      "ENSG00000174343.5 completed\n",
      "ENSG00000234492.3 completed\n",
      "ENSG00000151806.9 completed\n",
      "ENSG00000159674.7 completed\n",
      "ENSG00000243449.2 completed\n",
      "ENSG00000129187.10 completed\n",
      "ENSG00000168214.16 completed\n",
      "ENSG00000249780.1 completed\n",
      "ENSG00000246095.2 completed\n",
      "ENSG00000118804.7 completed\n",
      "ENSG00000138670.10 completed\n",
      "ENSG00000246090.2 completed\n",
      "ENSG00000164024.7 completed\n",
      "ENSG00000260244.1 completed\n",
      "ENSG00000138674.12 completed\n",
      "ENSG00000251600.1 completed\n",
      "ENSG00000129116.13 completed\n",
      "ENSG00000138658.11 completed\n",
      "ENSG00000251464.1 completed\n",
      "ENSG00000237350.1 completed\n",
      "ENSG00000229924.2 completed\n",
      "ENSG00000250163.1 completed\n",
      "ENSG00000232021.2 completed\n",
      "ENSG00000182308.5 completed\n",
      "ENSG00000164037.11 completed\n",
      "ENSG00000241279.1 completed\n",
      "ENSG00000250771.1 completed\n",
      "ENSG00000239735.1 completed\n",
      "ENSG00000145335.11 completed\n",
      "ENSG00000249592.1 completed\n",
      "ENSG00000250403.1 completed\n",
      "ENSG00000249960.1 completed\n",
      "ENSG00000248019.2 completed\n",
      "ENSG00000249465.1 completed\n",
      "ENSG00000130997.11 completed\n",
      "ENSG00000169435.9 completed\n",
      "ENSG00000181004.5 completed\n",
      "ENSG00000249464.1 completed\n",
      "ENSG00000163682.10 completed\n",
      "ENSG00000154447.10 completed\n",
      "ENSG00000248924.1 completed\n",
      "ENSG00000164032.7 completed\n",
      "ENSG00000173542.4 completed\n",
      "ENSG00000152583.7 completed\n",
      "ENSG00000109805.5 completed\n",
      "ENSG00000151552.7 completed\n",
      "ENSG00000151014.4 completed\n",
      "ENSG00000235413.3 completed\n",
      "ENSG00000205129.4 completed\n",
      "ENSG00000087269.10 completed\n",
      "ENSG00000071127.11 completed\n",
      "ENSG00000138646.4 completed\n",
      "ENSG00000164305.12 completed\n",
      "ENSG00000244021.2 completed\n",
      "ENSG00000145414.4 completed\n",
      "ENSG00000128040.6 completed\n",
      "ENSG00000078177.8 completed\n",
      "ENSG00000237868.1 completed\n",
      "ENSG00000109572.8 completed\n",
      "ENSG00000231841.1 completed\n",
      "ENSG00000145388.8 completed\n",
      "ENSG00000109390.7 completed\n",
      "ENSG00000197712.6 completed\n",
      "ENSG00000213480.3 completed\n",
      "ENSG00000237765.2 completed\n",
      "ENSG00000138764.9 completed\n",
      "ENSG00000168884.10 completed\n",
      "ENSG00000047315.10 completed\n",
      "ENSG00000248971.1 completed\n",
      "ENSG00000156136.5 completed\n",
      "ENSG00000163950.7 completed\n",
      "ENSG00000234801.2 completed\n",
      "ENSG00000248305.1 completed\n",
      "ENSG00000251160.1 completed\n",
      "ENSG00000248375.1 completed\n",
      "ENSG00000168826.11 completed\n",
      "ENSG00000118785.8 completed\n",
      "ENSG00000243894.1 completed\n",
      "ENSG00000251111.1 completed\n",
      "ENSG00000168538.10 completed\n",
      "ENSG00000163104.12 completed\n",
      "ENSG00000234036.3 completed\n",
      "ENSG00000138639.13 completed\n",
      "ENSG00000250826.1 completed\n",
      "ENSG00000170365.5 completed\n",
      "ENSG00000138801.4 completed\n",
      "ENSG00000241923.2 completed\n",
      "ENSG00000251411.1 completed\n",
      "ENSG00000245112.2 completed\n",
      "ENSG00000250050.1 completed\n",
      "Insufficient model output for SNP rs1959807, skipping.\n",
      "ENSG00000180613.6 completed\n",
      "ENSG00000250442.1 completed\n",
      "ENSG00000242882.1 completed\n",
      "ENSG00000138744.8 completed\n",
      "ENSG00000242262.1 completed\n",
      "ENSG00000164038.10 completed\n",
      "ENSG00000132466.12 completed\n",
      "ENSG00000138668.13 completed\n",
      "ENSG00000187653.10 completed\n",
      "ENSG00000124399.3 completed\n",
      "ENSG00000155016.12 completed\n",
      "ENSG00000248356.1 completed\n",
      "ENSG00000249635.1 completed\n",
      "ENSG00000163322.9 completed\n",
      "ENSG00000164120.9 completed\n",
      "ENSG00000183423.6 completed\n",
      "ENSG00000164039.10 completed\n",
      "ENSG00000138660.7 completed\n",
      "ENSG00000250966.1 completed\n",
      "ENSG00000145287.6 completed\n",
      "ENSG00000109445.6 completed\n",
      "ENSG00000255072.1 completed\n",
      "ENSG00000002587.5 completed\n",
      "ENSG00000244422.2 completed\n",
      "ENSG00000249311.2 completed\n",
      "ENSG00000248611.1 completed\n",
      "ENSG00000170185.5 completed\n",
      "ENSG00000213493.3 completed\n",
      "ENSG00000179010.10 completed\n",
      "ENSG00000232648.3 completed\n",
      "ENSG00000137460.4 completed\n",
      "ENSG00000125388.14 completed\n",
      "ENSG00000224097.3 completed\n",
      "ENSG00000186234.7 completed\n",
      "ENSG00000171497.4 completed\n",
      "ENSG00000180346.2 completed\n",
      "ENSG00000151726.8 completed\n",
      "ENSG00000245954.2 completed\n",
      "ENSG00000137463.4 completed\n",
      "ENSG00000121897.9 completed\n",
      "ENSG00000163297.12 completed\n",
      "ENSG00000145214.9 completed\n",
      "ENSG00000196526.6 completed\n",
      "ENSG00000138760.4 completed\n",
      "ENSG00000138758.7 completed\n",
      "ENSG00000188993.3 completed\n",
      "ENSG00000249096.1 completed\n",
      "ENSG00000007062.7 completed\n",
      "ENSG00000035720.3 completed\n",
      "ENSG00000248329.1 completed\n",
      "ENSG00000038002.4 completed\n",
      "ENSG00000169026.8 completed\n",
      "ENSG00000254531.1 completed\n",
      "ENSG00000234568.3 completed\n",
      "ENSG00000138767.8 completed\n",
      "ENSG00000163995.13 completed\n",
      "ENSG00000152990.9 completed\n",
      "ENSG00000248880.1 completed\n",
      "ENSG00000109814.7 completed\n",
      "ENSG00000170522.5 completed\n",
      "ENSG00000138768.9 completed\n",
      "ENSG00000169429.6 completed\n",
      "ENSG00000181826.5 completed\n",
      "ENSG00000240459.2 completed\n",
      "ENSG00000248110.1 completed\n",
      "ENSG00000109762.10 completed\n",
      "ENSG00000163631.11 completed\n",
      "ENSG00000227765.3 completed\n",
      "ENSG00000243175.1 completed\n",
      "ENSG00000213331.4 completed\n",
      "ENSG00000245322.2 completed\n",
      "ENSG00000068078.12 completed\n",
      "ENSG00000118816.5 completed\n",
      "ENSG00000249534.1 completed\n",
      "ENSG00000250075.1 completed\n",
      "ENSG00000250486.1 completed\n",
      "ENSG00000168564.5 completed\n",
      "ENSG00000251580.1 completed\n",
      "ENSG00000173966.7 completed\n",
      "ENSG00000137441.7 completed\n",
      "ENSG00000084092.6 completed\n",
      "ENSG00000241103.1 completed\n",
      "ENSG00000163293.7 completed\n",
      "ENSG00000243312.2 completed\n"
     ]
    }
   ],
   "source": [
    "# We will ramp up the process above, performing the analysis on a whole chromosome\n",
    "# I have selected chromosome 14 for our test\n",
    "chromosome = \"14\"\n",
    "chr14_genes = gene_expr[gene_expr['Chr'] == chromosome]['Gene_Symbol']\n",
    "\n",
    "# Initialize a list to store results for each SNP and gene as rows\n",
    "chr14_results_list = []\n",
    "\n",
    "for the_gene in chr4_genes:\n",
    "    gene_info = gene_expr[gene_expr['Gene_Symbol'] == the_gene]\n",
    "    snps_in_window = bim[(bim['chrom'] == chromosome) & \n",
    "                         (bim['pos'] >= int(gene_info['Coord'].iloc[0])-500000) & \n",
    "                         (bim['pos'] <= int(gene_info['Coord'].iloc[0])+500000)]\n",
    "    \n",
    "    # Get the SNP IDs from the filtered SNPs\n",
    "    snp_ids = snps_in_window['snp'].values\n",
    "\n",
    "    # Filter the genotype DataFrame for the SNPs in the window\n",
    "    genotype_data_window = geno_df.loc[:, snp_ids]\n",
    "    \n",
    "    # Filter both dataframes to only include matching IDs\n",
    "    geno_df_filtered = genotype_data_window.loc[matching_ids]\n",
    "    gene_expr_filtered = gene_expr_t.loc[matching_ids]\n",
    "\n",
    "    # Filter the gene expression data for the gene of interest\n",
    "    gene_expr_filtered = gene_expr_t.loc[matching_ids, the_gene]\n",
    "\n",
    "    # Merge the gene expression data with genotype data\n",
    "    merged_df = pd.concat([gene_expr_filtered, geno_df_filtered], axis=1)\n",
    "\n",
    "    # Rename the gene expression column for clarity\n",
    "    merged_df.rename(columns={the_gene: 'gene_expression'}, inplace=True)\n",
    "    \n",
    "    # Ensure gene expression values are numeric\n",
    "    y = pd.to_numeric(merged_df['gene_expression'], errors='coerce')\n",
    "\n",
    "    for snp in geno_df_filtered.columns:\n",
    "        X = pd.to_numeric(merged_df[snp], errors='coerce')  # Convert SNP data to numeric\n",
    "        X = sm.add_constant(X)  # Add an intercept for the linear regression\n",
    "    \n",
    "        # Drop rows with missing or invalid data (NaN)\n",
    "        valid_data = pd.concat([y, X], axis=1).dropna()\n",
    "    \n",
    "        # Run the regression only if valid data exists\n",
    "        if len(valid_data) > 0:\n",
    "            model = sm.OLS(valid_data['gene_expression'], valid_data.iloc[:, 1:]).fit()  # Use .iloc to access predictors\n",
    "\n",
    "            if len(model.pvalues) > 1:\n",
    "                # Append results for this SNP as a row in the results list\n",
    "                chr4_results_list = chr14_results_list.append({\n",
    "                    'Gene': the_gene,\n",
    "                    'SNP': snp,\n",
    "                    'p-value': model.pvalues.iloc[1],\n",
    "                    'beta': model.params.iloc[1],\n",
    "                    'r_squared': model.rsquared\n",
    "                })\n",
    "            else:\n",
    "                print(f\"Insufficient model output for SNP {snp}, skipping.\")\n",
    "                \n",
    "    print(the_gene + ' completed')\n",
    "\n",
    "# Convert the list of results into a DataFrame\n",
    "chr14_results_df = pd.DataFrame(chr14_results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b2185b42",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr14_results_df.to_csv('chr14_eqtl_data.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "76f64272",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>SNP</th>\n",
       "      <th>p-value</th>\n",
       "      <th>beta</th>\n",
       "      <th>r_squared</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs2799784</td>\n",
       "      <td>0.480146</td>\n",
       "      <td>-0.002291</td>\n",
       "      <td>1.458764e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs1959919</td>\n",
       "      <td>0.812529</td>\n",
       "      <td>-0.000658</td>\n",
       "      <td>1.646931e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs8008487</td>\n",
       "      <td>0.072117</td>\n",
       "      <td>-0.004474</td>\n",
       "      <td>9.425716e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs1959918</td>\n",
       "      <td>0.812529</td>\n",
       "      <td>-0.000658</td>\n",
       "      <td>1.646931e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000249212.1</td>\n",
       "      <td>rs1959928</td>\n",
       "      <td>0.012317</td>\n",
       "      <td>-0.009792</td>\n",
       "      <td>1.817704e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165509</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs10148007</td>\n",
       "      <td>0.595599</td>\n",
       "      <td>0.011221</td>\n",
       "      <td>8.245056e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165510</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs1349376</td>\n",
       "      <td>0.548497</td>\n",
       "      <td>0.012643</td>\n",
       "      <td>1.053664e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165511</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs8006417</td>\n",
       "      <td>0.824377</td>\n",
       "      <td>-0.004875</td>\n",
       "      <td>1.442003e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165512</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs8006648</td>\n",
       "      <td>0.995422</td>\n",
       "      <td>-0.000131</td>\n",
       "      <td>9.638331e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165513</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>rs2167239</td>\n",
       "      <td>0.883855</td>\n",
       "      <td>-0.003469</td>\n",
       "      <td>6.248916e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>165514 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene         SNP   p-value      beta     r_squared\n",
       "0       ENSG00000249212.1   rs2799784  0.480146 -0.002291  1.458764e-03\n",
       "1       ENSG00000249212.1   rs1959919  0.812529 -0.000658  1.646931e-04\n",
       "2       ENSG00000249212.1   rs8008487  0.072117 -0.004474  9.425716e-03\n",
       "3       ENSG00000249212.1   rs1959918  0.812529 -0.000658  1.646931e-04\n",
       "4       ENSG00000249212.1   rs1959928  0.012317 -0.009792  1.817704e-02\n",
       "...                   ...         ...       ...       ...           ...\n",
       "165509  ENSG00000243312.2  rs10148007  0.595599  0.011221  8.245056e-04\n",
       "165510  ENSG00000243312.2   rs1349376  0.548497  0.012643  1.053664e-03\n",
       "165511  ENSG00000243312.2   rs8006417  0.824377 -0.004875  1.442003e-04\n",
       "165512  ENSG00000243312.2   rs8006648  0.995422 -0.000131  9.638331e-08\n",
       "165513  ENSG00000243312.2   rs2167239  0.883855 -0.003469  6.248916e-05\n",
       "\n",
       "[165514 rows x 5 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr14_results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39f233cb",
   "metadata": {},
   "source": [
    "## Creating a PRS from the eQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "edb58967",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "cannot reindex on an axis with duplicate labels",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_168/994109592.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;31m# Compute PRS: sum of (genotype * beta) for each individual\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m     \u001b[0mprs_scores\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mthreshold\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mgenotype_data\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mbeta_values\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;31m# Store the PRS in a DataFrame for easy access\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/ops/common.py\u001b[0m in \u001b[0;36mnew_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     74\u001b[0m         \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem_from_zerodim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 76\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     77\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnew_method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/arraylike.py\u001b[0m in \u001b[0;36m__mul__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    200\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0munpack_zerodim_and_defer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"__mul__\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    201\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__mul__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 202\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_arith_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moperator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmul\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    203\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    204\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0munpack_zerodim_and_defer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"__rmul__\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_arith_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   7892\u001b[0m         \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mops\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmaybe_prepare_scalar_for_op\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7893\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7894\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_align_for_op\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7895\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7896\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merrstate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_align_for_op\u001b[0;34m(self, other, axis, flex, level)\u001b[0m\n\u001b[1;32m   8193\u001b[0m                     )\n\u001b[1;32m   8194\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 8195\u001b[0;31m             left, right = left.align(\n\u001b[0m\u001b[1;32m   8196\u001b[0m                 \u001b[0mright\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   8197\u001b[0m                 \u001b[0mjoin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"outer\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36malign\u001b[0;34m(self, other, join, axis, level, copy, fill_value, method, limit, fill_axis, broadcast_axis)\u001b[0m\n\u001b[1;32m  10439\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10440\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m> 10441\u001b[0;31m             left, _right, join_index = self._align_series(\n\u001b[0m\u001b[1;32m  10442\u001b[0m                 \u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10443\u001b[0m                 \u001b[0mjoin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_align_series\u001b[0;34m(self, other, join, axis, level, copy, fill_value, method, limit, fill_axis)\u001b[0m\n\u001b[1;32m  10582\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlidx\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10583\u001b[0m                 \u001b[0mbm_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_block_manager_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m> 10584\u001b[0;31m                 \u001b[0mfdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjoin_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlidx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbm_axis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m  10585\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10586\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcopy\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mfdata\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mreindex_indexer\u001b[0;34m(self, new_axis, indexer, axis, fill_value, allow_dups, copy, only_slice, use_na_proxy)\u001b[0m\n\u001b[1;32m    673\u001b[0m         \u001b[0;31m# some axes don't allow reindexing with dups\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    674\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mallow_dups\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 675\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_can_reindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    676\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    677\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_validate_can_reindex\u001b[0;34m(self, indexer)\u001b[0m\n\u001b[1;32m   4316\u001b[0m         \u001b[0;31m# trying to reindex on an axis with duplicates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4317\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_index_as_unique\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4318\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"cannot reindex on an axis with duplicate labels\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4319\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4320\u001b[0m     def reindex(\n",
      "\u001b[0;31mValueError\u001b[0m: cannot reindex on an axis with duplicate labels"
     ]
    }
   ],
   "source": [
    "# A potential clumping method:\n",
    "'''\n",
    "# Define the command as a list of strings\n",
    "command = [\n",
    "    \"plink\",\n",
    "    \"--bfile\", \"EUR.14.test${gene}\",\n",
    "    \"--clump-p1\", \"0.0001\",\n",
    "    \"--clump-r2\", \"0.1\",\n",
    "    \"--clump-kb\", \"250\",\n",
    "    \"--clump\", \"chr14_eqtl_data.txt\",\n",
    "    \"--clump-snp-field\", \"SNP\",\n",
    "    \"--clump-field\", \"P\",\n",
    "    \"--out\", \"EUR.14.test_clumped\"\n",
    "]\n",
    "\n",
    "# Run the PLINK command\n",
    "subprocess.run(command)\n",
    "'''\n",
    "\n",
    "# Instead I'm going to figure out which p-value works best for pruning\n",
    "# List of p-value thresholds\n",
    "thresholds = [1, 0.05, 0.01, 0.001, 0.00001]\n",
    "\n",
    "# Initialize a dictionary to store PRS scores\n",
    "prs_scores = {}\n",
    "\n",
    "# Loop over each threshold\n",
    "for threshold in thresholds:\n",
    "    # Filter SNPs based on the current threshold\n",
    "    snps_to_include = chr14_results_df[chr14_results_df['p-value'] <= threshold]\n",
    "    \n",
    "    # Retrieve genotype data for these SNPs\n",
    "    snp_ids = snps_to_include['SNP'].values\n",
    "    beta_values = snps_to_include.set_index('SNP')['beta']  # map SNPs to their beta values\n",
    "\n",
    "    # Select genotypes for these SNPs\n",
    "    genotype_data = geno_df.loc[:, snp_ids]\n",
    "\n",
    "    # Compute PRS: sum of (genotype * beta) for each individual\n",
    "    prs_scores[threshold] = (genotype_data * beta_values).sum(axis=1)\n",
    "\n",
    "# Store the PRS in a DataFrame for easy access\n",
    "prs_df = pd.DataFrame(prs_scores)\n",
    "prs_df.columns = [f\"PRS_{p}\" for p in thresholds]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "e12bb1b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TargetID</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Coord</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>...</th>\n",
       "      <th>NA20810</th>\n",
       "      <th>NA20811</th>\n",
       "      <th>NA20812</th>\n",
       "      <th>NA20813</th>\n",
       "      <th>NA20814</th>\n",
       "      <th>NA20815</th>\n",
       "      <th>NA20816</th>\n",
       "      <th>NA20819</th>\n",
       "      <th>NA20826</th>\n",
       "      <th>NA20828</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000152931.6</td>\n",
       "      <td>ENSG00000152931.6</td>\n",
       "      <td>5</td>\n",
       "      <td>59783540</td>\n",
       "      <td>0.101858</td>\n",
       "      <td>0.078110</td>\n",
       "      <td>0.048981</td>\n",
       "      <td>0.118597</td>\n",
       "      <td>0.004035</td>\n",
       "      <td>0.010925</td>\n",
       "      <td>...</td>\n",
       "      <td>0.088601</td>\n",
       "      <td>0.240010</td>\n",
       "      <td>0.137175</td>\n",
       "      <td>0.148494</td>\n",
       "      <td>0.038643</td>\n",
       "      <td>0.088509</td>\n",
       "      <td>0.029204</td>\n",
       "      <td>0.024423</td>\n",
       "      <td>0.044816</td>\n",
       "      <td>0.139186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000183696.9</td>\n",
       "      <td>ENSG00000183696.9</td>\n",
       "      <td>7</td>\n",
       "      <td>48128225</td>\n",
       "      <td>8.183805</td>\n",
       "      <td>5.686911</td>\n",
       "      <td>2.434653</td>\n",
       "      <td>3.830894</td>\n",
       "      <td>6.612288</td>\n",
       "      <td>4.709646</td>\n",
       "      <td>...</td>\n",
       "      <td>13.428205</td>\n",
       "      <td>6.094500</td>\n",
       "      <td>12.536000</td>\n",
       "      <td>2.217262</td>\n",
       "      <td>3.573394</td>\n",
       "      <td>7.583364</td>\n",
       "      <td>4.052882</td>\n",
       "      <td>1.570378</td>\n",
       "      <td>4.900372</td>\n",
       "      <td>6.737308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000139269.2</td>\n",
       "      <td>ENSG00000139269.2</td>\n",
       "      <td>12</td>\n",
       "      <td>57846106</td>\n",
       "      <td>1.199910</td>\n",
       "      <td>1.573572</td>\n",
       "      <td>0.521616</td>\n",
       "      <td>1.447225</td>\n",
       "      <td>3.565791</td>\n",
       "      <td>1.982681</td>\n",
       "      <td>...</td>\n",
       "      <td>3.225880</td>\n",
       "      <td>1.996067</td>\n",
       "      <td>2.854923</td>\n",
       "      <td>2.267343</td>\n",
       "      <td>1.331201</td>\n",
       "      <td>2.187895</td>\n",
       "      <td>1.004250</td>\n",
       "      <td>3.003316</td>\n",
       "      <td>1.984362</td>\n",
       "      <td>1.684954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000169129.8</td>\n",
       "      <td>ENSG00000169129.8</td>\n",
       "      <td>10</td>\n",
       "      <td>116164515</td>\n",
       "      <td>0.831940</td>\n",
       "      <td>0.069778</td>\n",
       "      <td>0.931086</td>\n",
       "      <td>0.620941</td>\n",
       "      <td>1.660668</td>\n",
       "      <td>0.570481</td>\n",
       "      <td>...</td>\n",
       "      <td>1.023381</td>\n",
       "      <td>1.127852</td>\n",
       "      <td>0.774409</td>\n",
       "      <td>1.495854</td>\n",
       "      <td>0.895342</td>\n",
       "      <td>1.513521</td>\n",
       "      <td>0.826377</td>\n",
       "      <td>1.021201</td>\n",
       "      <td>0.952502</td>\n",
       "      <td>0.740565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000134602.11</td>\n",
       "      <td>ENSG00000134602.11</td>\n",
       "      <td>X</td>\n",
       "      <td>131157293</td>\n",
       "      <td>27.646422</td>\n",
       "      <td>24.395572</td>\n",
       "      <td>16.445374</td>\n",
       "      <td>24.806650</td>\n",
       "      <td>25.113349</td>\n",
       "      <td>19.233988</td>\n",
       "      <td>...</td>\n",
       "      <td>25.079490</td>\n",
       "      <td>28.725528</td>\n",
       "      <td>24.450520</td>\n",
       "      <td>27.264069</td>\n",
       "      <td>26.912814</td>\n",
       "      <td>29.509210</td>\n",
       "      <td>26.462331</td>\n",
       "      <td>25.624009</td>\n",
       "      <td>25.707741</td>\n",
       "      <td>22.824957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23717</th>\n",
       "      <td>ENSG00000235472.1</td>\n",
       "      <td>ENSG00000235472.1</td>\n",
       "      <td>13</td>\n",
       "      <td>29172970</td>\n",
       "      <td>31.582832</td>\n",
       "      <td>34.071123</td>\n",
       "      <td>19.394365</td>\n",
       "      <td>37.523721</td>\n",
       "      <td>33.430473</td>\n",
       "      <td>37.844966</td>\n",
       "      <td>...</td>\n",
       "      <td>47.905182</td>\n",
       "      <td>33.224977</td>\n",
       "      <td>39.827675</td>\n",
       "      <td>27.096811</td>\n",
       "      <td>34.686342</td>\n",
       "      <td>37.386766</td>\n",
       "      <td>30.598840</td>\n",
       "      <td>33.516674</td>\n",
       "      <td>32.017940</td>\n",
       "      <td>38.341888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23718</th>\n",
       "      <td>ENSG00000114423.14</td>\n",
       "      <td>ENSG00000114423.14</td>\n",
       "      <td>3</td>\n",
       "      <td>105588396</td>\n",
       "      <td>14.054749</td>\n",
       "      <td>14.477899</td>\n",
       "      <td>11.584425</td>\n",
       "      <td>12.637956</td>\n",
       "      <td>12.015089</td>\n",
       "      <td>13.750655</td>\n",
       "      <td>...</td>\n",
       "      <td>11.723462</td>\n",
       "      <td>9.900372</td>\n",
       "      <td>10.473115</td>\n",
       "      <td>13.433413</td>\n",
       "      <td>15.832594</td>\n",
       "      <td>19.216176</td>\n",
       "      <td>10.213739</td>\n",
       "      <td>14.563192</td>\n",
       "      <td>15.637732</td>\n",
       "      <td>8.357117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23719</th>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>ENSG00000243312.2</td>\n",
       "      <td>4</td>\n",
       "      <td>87791344</td>\n",
       "      <td>1.112114</td>\n",
       "      <td>0.831797</td>\n",
       "      <td>0.253228</td>\n",
       "      <td>0.271568</td>\n",
       "      <td>0.486086</td>\n",
       "      <td>1.362640</td>\n",
       "      <td>...</td>\n",
       "      <td>1.168991</td>\n",
       "      <td>0.645389</td>\n",
       "      <td>0.819469</td>\n",
       "      <td>0.515448</td>\n",
       "      <td>0.463054</td>\n",
       "      <td>1.580658</td>\n",
       "      <td>0.701396</td>\n",
       "      <td>0.771233</td>\n",
       "      <td>0.857330</td>\n",
       "      <td>0.703369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23720</th>\n",
       "      <td>ENSG00000257337.1</td>\n",
       "      <td>ENSG00000257337.1</td>\n",
       "      <td>12</td>\n",
       "      <td>53448222</td>\n",
       "      <td>3.826396</td>\n",
       "      <td>6.045798</td>\n",
       "      <td>2.593872</td>\n",
       "      <td>4.447169</td>\n",
       "      <td>5.294657</td>\n",
       "      <td>4.106823</td>\n",
       "      <td>...</td>\n",
       "      <td>6.524398</td>\n",
       "      <td>6.017795</td>\n",
       "      <td>3.460273</td>\n",
       "      <td>3.905032</td>\n",
       "      <td>5.023161</td>\n",
       "      <td>5.333027</td>\n",
       "      <td>5.694370</td>\n",
       "      <td>8.142939</td>\n",
       "      <td>5.622043</td>\n",
       "      <td>6.026476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23721</th>\n",
       "      <td>ENSG00000177494.5</td>\n",
       "      <td>ENSG00000177494.5</td>\n",
       "      <td>3</td>\n",
       "      <td>111314290</td>\n",
       "      <td>1.183398</td>\n",
       "      <td>0.892619</td>\n",
       "      <td>0.917497</td>\n",
       "      <td>3.504894</td>\n",
       "      <td>1.168247</td>\n",
       "      <td>3.110953</td>\n",
       "      <td>...</td>\n",
       "      <td>1.936493</td>\n",
       "      <td>1.177972</td>\n",
       "      <td>2.673129</td>\n",
       "      <td>13.892711</td>\n",
       "      <td>2.763749</td>\n",
       "      <td>3.032246</td>\n",
       "      <td>4.558199</td>\n",
       "      <td>2.807851</td>\n",
       "      <td>2.021329</td>\n",
       "      <td>3.181096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23722 rows × 466 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TargetID         Gene_Symbol Chr      Coord    HG00096  \\\n",
       "0       ENSG00000152931.6   ENSG00000152931.6   5   59783540   0.101858   \n",
       "1       ENSG00000183696.9   ENSG00000183696.9   7   48128225   8.183805   \n",
       "2       ENSG00000139269.2   ENSG00000139269.2  12   57846106   1.199910   \n",
       "3       ENSG00000169129.8   ENSG00000169129.8  10  116164515   0.831940   \n",
       "4      ENSG00000134602.11  ENSG00000134602.11   X  131157293  27.646422   \n",
       "...                   ...                 ...  ..        ...        ...   \n",
       "23717   ENSG00000235472.1   ENSG00000235472.1  13   29172970  31.582832   \n",
       "23718  ENSG00000114423.14  ENSG00000114423.14   3  105588396  14.054749   \n",
       "23719   ENSG00000243312.2   ENSG00000243312.2   4   87791344   1.112114   \n",
       "23720   ENSG00000257337.1   ENSG00000257337.1  12   53448222   3.826396   \n",
       "23721   ENSG00000177494.5   ENSG00000177494.5   3  111314290   1.183398   \n",
       "\n",
       "         HG00097    HG00099    HG00100    HG00101    HG00102  ...    NA20810  \\\n",
       "0       0.078110   0.048981   0.118597   0.004035   0.010925  ...   0.088601   \n",
       "1       5.686911   2.434653   3.830894   6.612288   4.709646  ...  13.428205   \n",
       "2       1.573572   0.521616   1.447225   3.565791   1.982681  ...   3.225880   \n",
       "3       0.069778   0.931086   0.620941   1.660668   0.570481  ...   1.023381   \n",
       "4      24.395572  16.445374  24.806650  25.113349  19.233988  ...  25.079490   \n",
       "...          ...        ...        ...        ...        ...  ...        ...   \n",
       "23717  34.071123  19.394365  37.523721  33.430473  37.844966  ...  47.905182   \n",
       "23718  14.477899  11.584425  12.637956  12.015089  13.750655  ...  11.723462   \n",
       "23719   0.831797   0.253228   0.271568   0.486086   1.362640  ...   1.168991   \n",
       "23720   6.045798   2.593872   4.447169   5.294657   4.106823  ...   6.524398   \n",
       "23721   0.892619   0.917497   3.504894   1.168247   3.110953  ...   1.936493   \n",
       "\n",
       "         NA20811    NA20812    NA20813    NA20814    NA20815    NA20816  \\\n",
       "0       0.240010   0.137175   0.148494   0.038643   0.088509   0.029204   \n",
       "1       6.094500  12.536000   2.217262   3.573394   7.583364   4.052882   \n",
       "2       1.996067   2.854923   2.267343   1.331201   2.187895   1.004250   \n",
       "3       1.127852   0.774409   1.495854   0.895342   1.513521   0.826377   \n",
       "4      28.725528  24.450520  27.264069  26.912814  29.509210  26.462331   \n",
       "...          ...        ...        ...        ...        ...        ...   \n",
       "23717  33.224977  39.827675  27.096811  34.686342  37.386766  30.598840   \n",
       "23718   9.900372  10.473115  13.433413  15.832594  19.216176  10.213739   \n",
       "23719   0.645389   0.819469   0.515448   0.463054   1.580658   0.701396   \n",
       "23720   6.017795   3.460273   3.905032   5.023161   5.333027   5.694370   \n",
       "23721   1.177972   2.673129  13.892711   2.763749   3.032246   4.558199   \n",
       "\n",
       "         NA20819    NA20826    NA20828  \n",
       "0       0.024423   0.044816   0.139186  \n",
       "1       1.570378   4.900372   6.737308  \n",
       "2       3.003316   1.984362   1.684954  \n",
       "3       1.021201   0.952502   0.740565  \n",
       "4      25.624009  25.707741  22.824957  \n",
       "...          ...        ...        ...  \n",
       "23717  33.516674  32.017940  38.341888  \n",
       "23718  14.563192  15.637732   8.357117  \n",
       "23719   0.771233   0.857330   0.703369  \n",
       "23720   8.142939   5.622043   6.026476  \n",
       "23721   2.807851   2.021329   3.181096  \n",
       "\n",
       "[23722 rows x 466 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ebddbb9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70bd491d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "21bbbe92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: ./plink2: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!./plink2 --bfile EUR.14.test${gene} --clump-p1 $0.0001 --clump-r2 $0.1 --clump-kb 250 --clump $chr14_eqtl_data.txt --clump-snp-field SNP --clump-field P --out EUR.14.test_clumped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e3c1de51",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'plink'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_168/640417140.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# Run the PLINK command\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    503\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'stderr'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPIPE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    504\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 505\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    506\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    507\u001b[0m             \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask)\u001b[0m\n\u001b[1;32m    949\u001b[0m                             encoding=encoding, errors=errors)\n\u001b[1;32m    950\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 951\u001b[0;31m             self._execute_child(args, executable, preexec_fn, close_fds,\n\u001b[0m\u001b[1;32m    952\u001b[0m                                 \u001b[0mpass_fds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcwd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0menv\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    953\u001b[0m                                 \u001b[0mstartupinfo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcreationflags\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36m_execute_child\u001b[0;34m(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, gid, gids, uid, umask, start_new_session)\u001b[0m\n\u001b[1;32m   1819\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0merrno_num\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1820\u001b[0m                         \u001b[0merr_msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrerror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno_num\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1821\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mchild_exception_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno_num\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1822\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mchild_exception_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1823\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'plink'"
     ]
    }
   ],
   "source": [
    "# Define the command as a list of strings\n",
    "command = [\n",
    "    \"plink\",\n",
    "    \"--bfile\", \"EUR.14.test${gene}\",\n",
    "    \"--clump-p1\", \"0.0001\",\n",
    "    \"--clump-r2\", \"0.1\",\n",
    "    \"--clump-kb\", \"250\",\n",
    "    \"--clump\", \"chr14_eqtl_data.txt\",\n",
    "    \"--clump-snp-field\", \"SNP\",\n",
    "    \"--clump-field\", \"P\",\n",
    "    \"--out\", \"EUR.14.test_clumped\"\n",
    "]\n",
    "\n",
    "# Run the PLINK command\n",
    "subprocess.run(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "264c8e82",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
